{"cells":[{"metadata":{},"cell_type":"markdown","source":"Original post: https://www.kaggle.com/yyoshiaki/eda-scanpy-bioinformatics-standard-analysis\n\nBecause internet access should be disabled in the submission file, I uploaded the Scanpy & leidenalg packages so that we can use features created by scanpy in our submission."},{"metadata":{},"cell_type":"markdown","source":"# EDA by scanpy which is one of the most usful library for single-cell RNAseq analysis\n\n![https://scanpy.readthedocs.io/en/latest/_static/Scanpy_Logo_BrightFG.svg](https://scanpy.readthedocs.io/en/latest/_static/Scanpy_Logo_BrightFG.svg)\n\nScanpy is a very useful library designed for single-cell RNAseq analysis which deals with the massive gene expression profile of cell x gene. Now, Let's take a look into the dataset by common workflow for bioinformaticians. Though this dataset is not a single-cell RNAseq experiment, the pattern of data resembles it. Therefore, I introduce the standard analysis workflow of single-cell RNAseq.\n\nTutolials : https://scanpy.readthedocs.io/en/stable/index.html\n\nOriginal Article : Wolf, F., Angerer, P. & Theis, F. SCANPY: large-scale single-cell gene expression data analysis. Genome Biol 19, 15 (2018). https://doi.org/10.1186/s13059-017-1382-0"},{"metadata":{"trusted":true},"cell_type":"code","source":"import sys\nsys.path.append('../input/scanpy-related-packages/related_packages')\n\nimport numpy as np\nimport pandas as pd\nimport scanpy as sc\nfrom anndata import AnnData\nimport matplotlib.pyplot as plt\nfrom tqdm import tqdm\n\nimport leidenalg\nimport scanpy as sc\nfrom anndata import AnnData\nfrom sklearn import preprocessing\n\n\nsc.settings.verbosity = 3             # verbosity: errors (0), warnings (1), info (2), hints (3)\nsc.logging.print_header()\nsc.settings.set_figure_params(dpi=80, facecolor='white')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"def label_encoding(train: pd.DataFrame, test: pd.DataFrame, encode_cols):\n    n_train = len(train)\n    idx_train = train.index\n    idx_test = test.index\n    train = pd.concat([train, test], sort=False).reset_index(drop=True)\n    for f in encode_cols:\n        try:\n            lbl = preprocessing.LabelEncoder()\n            train[f] = lbl.fit_transform(list(train[f].values))\n        except:\n            print(f)\n    test = train[n_train:].reset_index(drop=True)\n    train = train[:n_train]\n    train.index = idx_train\n    test.index = idx_test\n    return train, test","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"scanpy can take anndata object. In this notebook, I handle cell viabilities as gene expression to make it easier."},{"metadata":{"_uuid":"d629ff2d2480ee46fbb7e2d37f6b5fab8052498a","_cell_guid":"79c7e3d0-c299-4dcb-8224-4455121ee9b0","trusted":true},"cell_type":"code","source":"train = pd.read_csv('/kaggle/input/lish-moa/train_features.csv', \n                    index_col=0)\ntest = pd.read_csv('/kaggle/input/lish-moa/test_features.csv',\n                  index_col=0)\n\ntrain_target = pd.read_csv(\"../input/lish-moa/train_targets_scored.csv\", \n                           index_col=0)\n\ntrain, test = label_encoding(train, test, ['cp_type', 'cp_dose'])\ntrain['dataset'] = 'train'\ntest['dataset'] = 'test'\n\ndf = pd.concat([train, test])\n\nlist_obs_col = ['cp_type', 'cp_time', 'cp_dose', 'dataset']\nlist_genes = [x for x in df.columns if x not in list_obs_col]\n\nadata = AnnData(df[list_genes], obs=df[list_obs_col])\nadata.obs = pd.concat([adata.obs, train_target.reindex(df.index)], \n                      axis=1)\nadata","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"Biological dataset is very noisy, so we sometimes uses only highly variable genes for UMAP and Defferntially expressed genes."},{"metadata":{"trusted":true},"cell_type":"code","source":"sc.pp.highly_variable_genes(adata, min_mean=0.0125, max_mean=3, min_disp=0.5)\nsc.pl.highly_variable_genes(adata)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"adata.raw = adata\nadata = adata[:, adata.var.highly_variable]","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"sc.tl.pca(adata, svd_solver='arpack')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"PCA can explain dataset well."},{"metadata":{"trusted":true},"cell_type":"code","source":"sc.pl.pca(adata, color='g-0', size=10)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"PCA variance ratio is used to determin the threshold of PCx."},{"metadata":{"trusted":true},"cell_type":"code","source":"sc.pl.pca_variance_ratio(adata, log=True)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"sc.pp.neighbors(adata, n_neighbors=10, n_pcs=40)\nsc.tl.umap(adata)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"adata.obs","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"You can easily plot gene expression and other features."},{"metadata":{"trusted":true},"cell_type":"code","source":"sc.pl.umap(adata, color=['g-0', 'dataset', '5-alpha_reductase_inhibitor',\n                        '11-beta-hsd1_inhibitor', '11-beta-hsd1_inhibitor', 'acat_inhibitor',\n                        'acetylcholine_receptor_agonist', 'acetylcholine_receptor_antagonist', \n                         'acetylcholinesterase_inhibitor'],\n          size=10)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"sc.tl.leiden(adata)\nsc.pl.umap(adata, color='leiden')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"scanpy can easily take feature genes for each leiden clusters."},{"metadata":{"trusted":true},"cell_type":"code","source":"sc.tl.rank_genes_groups(adata, 'leiden', method='t-test')\nsc.pl.rank_genes_groups(adata, n_genes=25, sharey=False)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"Let's see cluster 0 specific genes."},{"metadata":{"trusted":true},"cell_type":"code","source":"sc.pl.umap(adata, color=['g-37', 'c-26', 'g-369', 'leiden'],\n          size=10)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"sc.pl.rank_genes_groups_matrixplot(adata, groupby='leiden', n_genes=3)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"X_train = pd.concat([adata.obs.loc[train.index, ['cp_type', 'cp_time', 'cp_dose']].reset_index(drop=True), \n              pd.DataFrame(adata.obsm['X_pca'][adata.obs.loc[train.index].reset_index().index]),\n              pd.DataFrame(adata.obsm['X_umap'][adata.obs.loc[train.index].reset_index().index]),\n              pd.get_dummies(pd.DataFrame(adata.obs['leiden'][adata.obs.loc[train.index].reset_index().index])\n                            ).drop('leiden_0', axis=1).reset_index(drop=True)],\n         axis=1)\n\nX_test = pd.concat([adata.obs.loc[test.index, ['cp_type', 'cp_time', 'cp_dose']].reset_index(drop=True), \n              pd.DataFrame(adata.obsm['X_pca'][adata.obs.loc[test.index].reset_index().index]),\n              pd.DataFrame(adata.obsm['X_umap'][adata.obs.loc[test.index].reset_index().index]),\n              pd.get_dummies(pd.DataFrame(adata.obs['leiden'][adata.obs.loc[test.index].reset_index().index])\n                            ).drop('leiden_0', axis=1).reset_index(drop=True)],\n             axis=1)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"from sklearn.linear_model import LogisticRegression\n\n\ndef lr(X_train, y_train, X_test):\n    reg = LogisticRegression().fit(X_train, y_train)\n    return reg.predict_proba(X_test)[:,1]\n\nlist_pred = []\nfor c in train_target.columns:\n    list_pred.append(lr(X_train, train_target[c], X_test))\ndf_pred = pd.DataFrame(list_pred).T\ndf_pred.index = test.index\ndf_pred.columns = train_target.columns\ndf_pred.to_csv('submission.csv')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"adata.obs.loc[df_pred.index, df_pred.columns] = df_pred","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"adata[adata.obs['dataset'] == 'train']","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"Check the predicted score in a sample population."},{"metadata":{"trusted":true},"cell_type":"code","source":"sc.pl.umap(adata[adata.obs['dataset'] == 'train'], color=['5-alpha_reductase_inhibitor',\n                        '11-beta-hsd1_inhibitor', '11-beta-hsd1_inhibitor', 'acat_inhibitor',\n                        'acetylcholine_receptor_agonist', 'acetylcholine_receptor_antagonist', \n                         'acetylcholinesterase_inhibitor'],\n          size=10)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"sc.pl.umap(adata[adata.obs['dataset'] == 'test'], color=['5-alpha_reductase_inhibitor',\n                        '11-beta-hsd1_inhibitor', '11-beta-hsd1_inhibitor', 'acat_inhibitor',\n                        'acetylcholine_receptor_agonist', 'acetylcholine_receptor_antagonist', \n                         'acetylcholinesterase_inhibitor'],\n          size=40)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}