{"cells":[{"metadata":{},"cell_type":"markdown","source":"# Mechanisms of Action (MoA) Prediction","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"The Connectivity Map, a project within the Broad Institute of MIT and Harvard, together with the Laboratory for Innovation Science at Harvard (LISH), presents this challenge with the goal of advancing drug development through improvements to MoA prediction algorithms.\n\nWhat is the Mechanism of Action (MoA) of a drug? And why is it important?\n\nIn the past, scientists derived drugs from natural products or were inspired by traditional remedies. Very common drugs, such as paracetamol, known in the US as acetaminophen, were put into clinical use decades before the biological mechanisms driving their pharmacological activities were understood. Today, with the advent of more powerful technologies, drug discovery has changed from the serendipitous approaches of the past to a more targeted model based on an understanding of the underlying biological mechanism of a disease. In this new framework, scientists seek to identify a protein target associated with a disease and develop a molecule that can modulate that protein target. As a shorthand to describe the biological activity of a given molecule, scientists assign a label referred to as mechanism-of-action or MoA for short.\n\nHow do we determine the MoAs of a new drug?\n\nOne approach is to treat a sample of human cells with the drug and then analyze the cellular responses with algorithms that search for similarity to known patterns in large genomic databases, such as libraries of gene expression or cell viability patterns of drugs with known MoAs.\n\nIn this competition, you will have access to a unique dataset that combines gene expression and cell viability data. The data is based on a new technology that measures simultaneously (within the same samples) human cells’ responses to drugs in a pool of 100 different cell types (thus solving the problem of identifying ex-ante, which cell types are better suited for a given drug). In addition, you will have access to MoA annotations for more than 5,000 drugs in this dataset.\n\nAs is customary, the dataset has been split into testing and training subsets. Hence, your task is to use the training dataset to develop an algorithm that automatically labels each case in the test set as one or more MoA classes. Note that since drugs can have multiple MoA annotations, the task is formally a multi-label classification problem.\n\nHow to evaluate the accuracy of a solution?\n\nBased on the MoA annotations, the accuracy of solutions will be evaluated on the average value of the logarithmic loss function applied to each drug-MoA annotation pair.\n\nIf successful, you’ll help to develop an algorithm to predict a compound’s MoA given its cellular signature, thus helping scientists advance the drug discovery process.","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"![](https://images.financialexpress.com/2020/01/drugs-1.jpg?w=1200&h=800&imflag=true)","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"**Data Description**\n\nIn this competition, you will be predicting multiple targets of the Mechanism of Action (MoA) response(s) of different samples (sig_id), given various inputs such as gene expression data and cell viability data.\n\n***Two notes:***\n\n* the training data has an additional (optional) set of MoA labels that are not included in the test data and not used for scoring.\n* the re-run dataset has approximately 4x the number of examples seen in the Public test.\n\n**Files**\n* **train_features.csv** - Features for the training set. Features g- signify gene expression data, and c- signify cell viability data. cp_type indicates samples treated with a compound (cp_vehicle) or with a control perturbation (ctrl_vehicle); control perturbations have no MoAs; cp_time and cp_dose indicate treatment duration (24, 48, 72 hours) and dose (high or low).\n* **train_targets_scored.csv** - The binary MoA targets that are scored.\n* **train_targets_nonscored.csv** - Additional (optional) binary MoA responses for the training data. These are not predicted nor scored.\n* **test_features.csv** - Features for the test data. You must predict the probability of each scored MoA for each row in the test data.\n* **sample_submission.csv** - A submission file in the correct format.","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"# Load the Packages","execution_count":null},{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true},"cell_type":"code","source":"# This Python 3 environment comes with many helpful analytics libraries installed\n# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n# For example, here's several helpful packages to load\n\nimport numpy as np # linear algebra\nimport pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\nimport seaborn as sns\nimport matplotlib.pyplot as plt\n\n# Input data files are available in the read-only \"../input/\" directory\n# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n\nimport os\nfor dirname, _, filenames in os.walk('/kaggle/input'):\n    for filename in filenames:\n        print(os.path.join(dirname, filename))\n\n# You can write up to 5GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using \"Save & Run All\" \n# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Load the Data","execution_count":null},{"metadata":{"_uuid":"d629ff2d2480ee46fbb7e2d37f6b5fab8052498a","_cell_guid":"79c7e3d0-c299-4dcb-8224-4455121ee9b0","trusted":true},"cell_type":"code","source":"df_Train = pd.read_csv('../input/lish-moa/train_features.csv')\ndf_Test = pd.read_csv('../input/lish-moa/test_features.csv')\ndf_Train_scored = pd.read_csv('../input/lish-moa/train_targets_scored.csv')\ndf_Train_unscored = pd.read_csv('../input/lish-moa/train_targets_nonscored.csv')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Exploratory Data Analysis","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"#To find the head of the Data\ndf_Train.head()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"df_Train_scored.head()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"#Columns List\ndf_Train.columns","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"#Columns List\ndf_Train_scored.columns","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"#Information of the Dataset Continuous Values\ndf_Train.describe()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"#Information of the Dataset Values\ndf_Train.info()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"#Information of the Dataset Values\ndf_Train_scored.info()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"#Shape of the Train and Test Data\nprint('Shape of Train Data: ', df_Train.shape)\nprint('Shape of Train Scored Data: ', df_Train_scored.shape)\nprint('Shape of Train Unscored Data: ', df_Train_unscored.shape)\nprint('Shape of Test Data: ', df_Test.shape)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### Missing Values","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"#Null values in the Train Dataset\nprint('Null values in Train Data: \\n', df_Train.isnull().sum())","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"#Null Values in the Test Dataset\nprint('Null Values in Test Data: \\n', df_Test.isnull().sum())","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# Data Insight and Visualization","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"## Correlation\n\nFinding the Correlation of the Target Column with the other Columns","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"**Correlation**\n\nTo check for the correlation of the variable and the graph dependence or association is any statistical relationship, whether causal or not, between two random variables or bivariate data. Correlation is any of a broad class of statistical relationships involving dependence, though in common usage it most often refers to how close two variables are to having a linear relationship with each other","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"corrmat = df_Train.corr()\nf, ax = plt.subplots(figsize=(14,14))\nsns.heatmap(corrmat, square=True, vmax=.8)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"corrmat = df_Train_scored.corr()\nf, ax = plt.subplots(figsize=(14,14))\nsns.heatmap(corrmat, square=True, vmax=.8)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"corrmat = df_Train_unscored.corr()\nf, ax = plt.subplots(figsize=(14,14))\nsns.heatmap(corrmat, square=True, vmax=.8)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}