{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"<div style=\"color:white;\n           display:fill;\n           border-radius:0px;\n           background-color:#004c6d;\n           font-size:110%;\n           font-family:Verdana;\n           letter-spacing:1px\">\n\n<p style=\"padding: 16px;\n              color:white;overflow:hidden;margin:0;font-size:100%;text-align:center\"> \n    <span style=\"font-size:30px;\">\n    <b>\nPrincipal Component Bacterial Genetic Analysis\n    </b>\n        </span>\n</p>\n</div>\n\n# <span style=\"color:#004c6d\"> Summary </span>\nOur task is to classify 10 different bacteria species using oligomer sampling methods outlined in Wood et al (2020). This method of genomic analysis extracts 10-mer snippets of DNA to find the frequency of nucleotide base counts in the sequence (= an oligomer, eg. AxTxGxCx) with machine learning to rapidly identify bacteria species.  \n\n\"Bacterial antibiotic resistance is becoming a significant health threat, and rapid\nidentification of antibiotic-resistant bacteria is essential to save lives and reduce the\nspread of antibiotic resistance. This paper analyzes the ability of machine learning\nalgorithms (MLAs) to process data from a novel spectroscopic diagnostic device to\nidentify antibiotic-resistant genes and bacterial species by comparison to available\nbacterial DNA sequences. Simulation results show that the algorithms attain from 92%\naccuracy (for genes) up to 99% accuracy (for species). This novel approach identifies\ngenes and species by optically reading the percentage of A, C, G, T bases in 1000s\nof short 10-base DNA oligomers instead of relying on conventional DNA sequencing\nin which the sequence of bases in long oligomers provides genetic information.\nThe identification algorithms are robust in the presence of simulated random genetic\nmutations and simulated random experimental errors. Thus, these algorithms can be\nused to identify bacterial species, to reveal antibiotic resistance genes, and to perform\nother genomic analyses. Some MLAs evaluated here are shown to be better than\nothers at accurate gene identification and avoidance of false negative identification of\nantibiotic resistance.\" (Wood, 2020)\n\nThe speed and accuracy of new genomic sequencing techniques provides cost effective and scalable becterial testing capabilities for scientists and has a broad range of use in society. The use of machine learning in research supports the advancement of science and the highlights the scope of knowledge that can be aquired from less information by using technology.\n\n**Key players** \n\n* Bacteria Species Samples\n* Oligomer Genetic Sequence Frequency\n<div style=\"color:white;\n           display:fill;\n           border-radius:0px;\n           background-color:#004c6d;\n           font-size:110%;\n           font-family:Verdana;\n           letter-spacing:1px\">\n\n# <span style=\"color:#004c6d\"> Stage 1:  Ask </span>\n#### \"The scientist is not a person who gives the right answers, he is one who asks the right questions.\"- Claude Lévi-Strauss\n\n### <span style=\"color:#004c6d\"> 1.1 Analysis Task </span>  \n\nTask: Machine learning tasks are usually split into three categories; supervised, unsupervised and reinforcement. For this competition, our task is supervised learning.\n\n### <span style=\"color:#004c6d\"> 1.2 Dataset used: </span>  \n\nWe are given a training and testing data set in csv format.\n\nTraining: the training data consists of 199,999 rows of bacterial samples, with 285 columns describing oligomer presence. We are have been provided the target bacteria species for each row.\n\nTesting: The test data set consists of 100,000 bacteria samples with the same attributes as the training data. However, the test data has does not have listed bacteria species, and will rely on the machine learning models to predict bacteria type based on oligomer frequency. We will be running and tuning multiple mlm models to create predictions for the Kaggle competition.\n\n### <span style=\"color:#004c6d\"> 1.3 Data Organization and verification: </span>\n\nThe data source used for this project is the tabular-playground-series-feb-2022 dataset on Kaggle. This dataset contains data for 399,000 bacteria samples. Each row represents one sample, the columns describe different oligomer permutations (ex. AxTxGxCx).\n\nThe two CSV documents- the training set (train.csv) and test set (test.csv) provide a similar distribution of bacterial samples, with the training set available to train machine learning models, and the test set to test model accuracy. \nThe data is considered wide since each row is one subject. Every bacterial sample has unique oligomer frequencies as columnarly listed attributes. \n\nDue to the size of the training set, I sorted and filtered tables creating Pivot Tables in Excel. \nI was able to verify attributes and observations.\n\n| Variable | Description |\n| --- | --- |   \n| row_id | Bacterial Sample Number |\n| A0T0G0C10 | Oligomer Frequency in Genome, a Permutation of AxTxGxCx |\n| ...x286 | ... |\n| A10T0G0C0 | Oligomer Frequency in Genome, a Permutation of AxTxGxCx |\n| target | Bacteria Species |\n\n**Task 1:** Choose data- For this analysis objective, I am going to work with the test and train datasets to conduct descriptive statistics and run machine learning models.\n\n**Task 2:** Know your audience- I am performing analysis for the Kaggle competition, and the Kaggle community. Kaggler's have a working understanding of data analytics and science. I need to tailor my results to be useful to them.\n\n**Task 3:** Aims to address objective- In the aims of answering this analytics objective, I decided analysis should identify genetic sequences with a strong classification effect on bacterial species and conduct descriptive statistics, as well as fit and tune multiple machine learning models. The oligomer sequences with the strongest effect were: A5T4G0C1, A5T5G0C0, A0T0G5C5, A0T0G7C3, A5T4G0C1, A5T5G0C0, A0T0G5C5, A0T0G7C3, A1T2G6C1, A1T2G7C0, A5T4G0C1, and A5T5G0C0.\n\n\n# <span style=\"color:#004c6d\"> Stage 2:  Prepare </span>     \n\n#### “It is a capital mistake to theorize before one has data.” -Sherlock Holmes\n\n**Task 1: Data source validation-** Verifying the metadata of the dataset confirmed it as open-source. The owner has dedicated the work to the public domain by waiving all of his or her rights to the work worldwide under copyright law, including all related and neighboring rights, to the extent allowed by law. We can copy, modify, and distribute the work without asking permission. \n\nWith information about only 399,000 bacterial samples, with no sampling methodology, there are limitations that could potentially be encountered due to sampling bias. This dataset is not representative of an entire population, and our model could be considered an operational approach to bacterial species identification. \n\n**Task 2: Data exploration-** I needed to prepare my workspace and load the packages that I would be using in my analysis. I uploaded the files to the Jupyter notebook hub, and made sure they were in a usable format. I used the describe.data function on each dataset to find general trends and irregularities in the data. In terms of Python syntax, I did not have a problem with the original column names or data distribution, so I kept them the same.\n\nI used Python and Jupyter notebooks for the competition. Due to the amount of data, needing to visualize and model data, and sharing my results through a markdown, I am using Python to run my analysis.","metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19"}},{"cell_type":"code","source":"# import data analysis libraries\nimport pandas as pd\nimport numpy as np\nimport random as rnd\nfrom pandas import DataFrame\nimport datetime\n\n# import data visualization libraries\nimport seaborn as sns\nimport matplotlib.pyplot as plt\n%matplotlib inline\nimport matplotlib.colors as mpl\nimport plotly.express as px\nimport plotly.graph_objects as go\nfrom plotly.subplots import make_subplots\nfrom plotly.offline import init_notebook_mode\n\n# import statistics packages\nfrom scipy.stats import skew, norm\nfrom scipy.stats import boxcox\nfrom scipy.special import boxcox1p\nfrom scipy.stats import boxcox_normmax\nimport statsmodels.api as sm \nfrom statsmodels.formula.api import ols\n\n# import machine learning models\nfrom sklearn.linear_model import LogisticRegression\nfrom sklearn.svm import SVC, LinearSVC\nfrom sklearn.ensemble import RandomForestClassifier, RandomForestRegressor, GradientBoostingRegressor, AdaBoostRegressor, BaggingRegressor\nfrom sklearn.ensemble import ExtraTreesClassifier\nfrom sklearn.kernel_ridge import KernelRidge\nfrom sklearn.neighbors import KNeighborsClassifier\nfrom sklearn.naive_bayes import GaussianNB\nfrom sklearn.linear_model import Perceptron\nfrom sklearn.linear_model import SGDClassifier\nfrom sklearn.tree import DecisionTreeClassifier\nfrom sklearn.linear_model import Ridge, RidgeCV\nfrom sklearn.linear_model import ElasticNet, ElasticNetCV\nfrom sklearn.svm import SVR\n!pip install mlxtend\nfrom mlxtend.regressor import StackingCVRegressor\n!pip install lightgbm\nfrom lightgbm import LGBMRegressor\n!pip install xgboost\nfrom xgboost import XGBRegressor\nfrom sklearn.pipeline import make_pipeline\nimport lightgbm as lgb\nfrom sklearn.preprocessing import StandardScaler, LabelEncoder\nfrom sklearn.decomposition import PCA\nfrom sklearn.model_selection import train_test_split, cross_val_score, StratifiedKFold, KFold\nfrom sklearn.model_selection import GridSearchCV\nfrom sklearn.metrics import accuracy_score, roc_curve, roc_auc_score, auc, classification_report, mean_squared_error\nfrom sklearn.preprocessing import OneHotEncoder\nfrom sklearn.preprocessing import LabelEncoder\nfrom sklearn.pipeline import make_pipeline\nfrom sklearn.preprocessing import scale\nfrom sklearn.preprocessing import StandardScaler\nfrom sklearn.preprocessing import RobustScaler\nfrom sklearn.decomposition import PCA\nfrom sklearn.preprocessing import StandardScaler, LabelEncoder\nfrom sklearn.cluster import KMeans\nfrom math import factorial\n\n# Stats\nfrom scipy import stats\nfrom scipy.stats import skew, norm\nfrom scipy.special import boxcox1p\nfrom scipy.stats import boxcox_normmax\n\npd.set_option('display.max_columns', None)\n\n# Ignore useless warnings\nimport warnings\nwarnings.filterwarnings(action=\"ignore\")\npd.options.display.max_seq_items = 8000\npd.options.display.max_rows = 8000\n\n# Load data\ntrain = pd.read_csv('../input/tabular-playground-series-feb-2022/train.csv', index_col=0)\ntest = pd.read_csv('../input/tabular-playground-series-feb-2022/test.csv', index_col=0)\nsub = pd.read_csv('../input/tabular-playground-series-feb-2022/sample_submission.csv')","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# <span style=\"color:#004c6d\"> Stage 3:  Process </span>\n\n#### \"No data is clean, but most is useful\" - Dean Abbott\n\n**Task 1:** Clean and process data- With a single bacterial sample serving as a unit of replication and the oligomer features already listed, I didn't need to further create or manipulate the metrics for this dataset. The dataset had no missing values. I removed duplicate values, identified the number of bacteria species present, and transformed the response varibles (oligomer frequency) to integer format. Later in the modeling process, I will set the dataset as factors and addressed asymmetry in the data source.\n\n### <span style=\"color:#004c6d\"> 3.1 Find number of missing values and duplicates </span>","metadata":{}},{"cell_type":"code","source":"# Load data success?\nprint(\"Train dataset has {} rows with {} variables each.\".format(*train.shape))\nprint(\"Test dataset has {} rows with {} variables each.\".format(*test.shape))","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:17:49.530921Z","iopub.execute_input":"2022-02-24T22:17:49.531374Z","iopub.status.idle":"2022-02-24T22:17:49.542819Z","shell.execute_reply.started":"2022-02-24T22:17:49.531322Z","shell.execute_reply":"2022-02-24T22:17:49.541639Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"#Count the number of missing and duplicate values\nprint('Train dataset has {} missing values and {} duplicate values'\\\n      .format(train.isna().sum().sum(), train.duplicated().sum()))\nprint('Test dataset has {} missing values and {} duplicates values'\\\n      .format(test.isna().sum().sum(), test.duplicated().sum()))\n\n#Drop duplicate values\ntrain_dropped=train.drop_duplicates() \nprint('Dropping Duplicate values:\\nNew Train Shape: {}'.format(train_dropped.shape))","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:17:49.544619Z","iopub.execute_input":"2022-02-24T22:17:49.544952Z","iopub.status.idle":"2022-02-24T22:17:54.869447Z","shell.execute_reply.started":"2022-02-24T22:17:49.544901Z","shell.execute_reply":"2022-02-24T22:17:54.868426Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### <span style=\"color:#004c6d\"> 3.2 Bacterial Species Data Exploration </span>","metadata":{}},{"cell_type":"code","source":"# identify bacterial species\ntrain_dropped.groupby('target').describe()","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:17:54.872333Z","iopub.execute_input":"2022-02-24T22:17:54.872899Z","iopub.status.idle":"2022-02-24T22:18:04.806208Z","shell.execute_reply.started":"2022-02-24T22:17:54.872841Z","shell.execute_reply":"2022-02-24T22:18:04.804775Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Bacteria Species Dataset Composition\nbact=train_dropped.target.value_counts(normalize=True).reset_index()\nbact.target=bact.target.mul(100).sort_values(ascending=False)\nbact.describe","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:18:04.807742Z","iopub.execute_input":"2022-02-24T22:18:04.808319Z","iopub.status.idle":"2022-02-24T22:18:04.843443Z","shell.execute_reply.started":"2022-02-24T22:18:04.808277Z","shell.execute_reply":"2022-02-24T22:18:04.842576Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"In the dataset, bacteria species are represented around 10%, at similar frequencies. There is no asymmetry from sampling bias that needs to be controlled.","metadata":{}},{"cell_type":"markdown","source":"### <span style=\"color:#004c6d\"> 3.2 Converting Frequecy Values to Integer Format </span>\nThe frequency values for bacterial oligomer frequency are in the format of a float point number, described in the data source. To convert these values to integer values, the methods are described in Wood et al. (2020). The authors list a data treatment process where frequency integers were divided by 1,000,000 and a specific constant, called the bias, was subtracted. I algebraically transformed values back to their integer format.","metadata":{}},{"cell_type":"code","source":"train_dropped_integers = train_dropped.copy()\ntest_integer = test.copy()\n\nelements = [e for e in train_dropped_integers.columns if e != 'row_id' and e != 'target']\n\n# Convert the 10 bacteria names to the integers 0 .. 9\nle = LabelEncoder()\ntrain_dropped_integers['target_num'] = le.fit_transform(train_dropped_integers.target)\n\ndef bias(w, x, y, z):\n    return factorial(10) / (factorial(w) * factorial(x) * factorial(y) * factorial(z) * 4**10)\n\ndef bias_of(s):\n    w = int(s[1:s.index('T')])\n    x = int(s[s.index('T')+1:s.index('G')])\n    y = int(s[s.index('G')+1:s.index('C')])\n    z = int(s[s.index('C')+1:])\n    return factorial(10) / (factorial(w) * factorial(x) * factorial(y) * factorial(z) * 4**10)\n\ntrain_integer = pd.DataFrame({col: ((train_dropped_integers[col] + bias_of(col)) * 1000000).round().astype(int)\n                        for col in elements})\ntest_integer = pd.DataFrame({col: ((test[col] + bias_of(col)) * 1000000).round().astype(int)\n                       for col in elements})\ntrain_integer","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:18:04.844855Z","iopub.execute_input":"2022-02-24T22:18:04.845709Z","iopub.status.idle":"2022-02-24T22:18:06.437537Z","shell.execute_reply.started":"2022-02-24T22:18:04.845673Z","shell.execute_reply":"2022-02-24T22:18:06.436578Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# <span style=\"color:#004c6d\"> Stage 4-5: Analyze and Figures   </span>\n#### \"If you torture the data long enough, it will confess.\" - Ronald Coase\n### <span style=\"color:#004c6d\"> 4.1 Exploratory correlation   </span> \n\nFor this section of descriptive statistics, I am investigating the variables with the highest rough correlation with bacteria species to see if these variables have a statistically significant relationship.","metadata":{}},{"cell_type":"code","source":"cor=train_integer.corr()    \ncor.style.background_gradient(cmap='plasma')","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:18:06.438987Z","iopub.execute_input":"2022-02-24T22:18:06.43927Z","iopub.status.idle":"2022-02-24T22:18:40.809433Z","shell.execute_reply.started":"2022-02-24T22:18:06.439226Z","shell.execute_reply":"2022-02-24T22:18:40.808239Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Figure 1. Correlation of bacterial oligomer frequecies in bacterial samples (n = 5.7e7). Oligomer genetic sequence is listed in correlation on x and y axes, with highly grouped values indicated by brighter colors.","metadata":{}},{"cell_type":"markdown","source":"### <span style=\"color:#004c6d\"> 4.2 Oligomer Correlation </span> \nBelow is a list of the highest 10 oligomer pairs correlated per species.","metadata":{}},{"cell_type":"code","source":"for i in train_dropped.target.unique():\n    cor_df=train_dropped[train_dropped.target==i]\n    cor=cor_df.corr()  \n    c = cor.abs().unstack().drop_duplicates().reset_index()\n    c = c.rename(columns={'level_0': 'Oligomer 1', 'level_1': 'Oligomer 2', 0: 'Correlation'})\n    c = c.query('.75 <= Correlation < 1').sort_values(by = 'Correlation', ascending = False).reset_index(drop=True)\n    display(c.iloc[:10,:].style.background_gradient(cmap='plasma').set_caption('Strongest oligomer correlation: {}'.format(i.replace('_', ' '))))","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:18:40.811018Z","iopub.execute_input":"2022-02-24T22:18:40.811329Z","iopub.status.idle":"2022-02-24T22:19:09.366751Z","shell.execute_reply.started":"2022-02-24T22:18:40.811293Z","shell.execute_reply":"2022-02-24T22:19:09.36527Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Based in the descriptive statistics, different oligomer types are more strongly correlated with different bacteral species. There are some repeating oligomer genotypes (ex. A0T0G5C5 and A0T0G7C3 in *Streptococcus pyogenes* and *Enterococcus hirae*). Bacterial species are subdivided to groups with a simmilar genetic structure.\nThis table summarize the results :\n\n| Rank | Species | Oligomer 1 | Oligomer 2 | Correlation |\n| --- | --- | --- | --- | --- |\n| 1 | *Streptococcus pneumoniae* | A0T0G8C2 | A1T0G0C9 | 0.995343|\n| 2 | *Campylobacter jejuni* | A0T0G4C6 | A0T0G6C4 | 0.994300 |\n| 3 | *Staphylococcus aureus* | A0T0G3C7 | A0T1G1C8 | 0.991636 |\n| 4 | *Streptococcus pyogenes* | A0T0G5C5 | A0T0G7C3 | 0.981113 |\n| 5 | *Enterococcus hirae* | A0T0G5C5 | A0T0G7C3 | 0.980968 |\n| 6 | *Klebsiella pneumoniae* | A5T4G0C1 | A5T5G0C0 | 0.956983 |\n| 7 | *Escherichia fergusonii* | A5T4G0C1 | A5T5G0C0 | 0.916947 |\n| 8 | *Salmonella enterica* | A5T4G0C1 | A5T5G0C0 | 0.916334 |\n| 9 | *Escherichia coli* | A5T4G0C1 | A5T5G0C0 | 0.904267 |\n| 10 | *Bacteroides fragilis* | A1T2G6C1 | A1T2G7C0 | 0.870281 |","metadata":{}},{"cell_type":"markdown","source":"### <span style=\"color:#004c6d\"> 4.3 Overall Oligomer Correlation </span> ","metadata":{}},{"cell_type":"code","source":"#Strongest Overall Oligomer Correlation\nc = cor.abs().unstack().drop_duplicates().reset_index()\nc = c.rename(columns={'level_0': 'Oligomer 1', 'level_1': 'Oligomer 2', 0: 'Correlation'})\nc = c.query('.85 <= Correlation < 1').sort_values(by = 'Correlation', ascending = False).reset_index(drop=True)\nc.style.background_gradient(cmap='plasma')","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:19:09.368463Z","iopub.execute_input":"2022-02-24T22:19:09.368811Z","iopub.status.idle":"2022-02-24T22:19:09.406742Z","shell.execute_reply.started":"2022-02-24T22:19:09.368766Z","shell.execute_reply":"2022-02-24T22:19:09.405718Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### <span style=\"color:#004c6d\"> 5.1 Figures: Species Type vs. Oligomer Frequency </span> ","metadata":{}},{"cell_type":"code","source":"# A0T0G5C5 bacterial species composition\nsns.set_context(\"notebook\", font_scale=1.5)\nsns.set_style(\"white\")\nsns.set_color_codes(palette='deep')\nf, ax = plt.subplots(figsize=(8, 6))\nsns.barplot(x = \"A0T0G5C5\", y = \"target\", palette= 'plasma', data = train_dropped)\nplt.tick_params(axis='x', labelsize=15, rotation=-45)\nplt.tick_params(axis='y', labelsize=15, rotation=0)\nplt.xlabel(\"Oligomer Frequency\")\nplt.ylabel(\"Bacterial Species\")","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:19:09.410676Z","iopub.execute_input":"2022-02-24T22:19:09.410947Z","iopub.status.idle":"2022-02-24T22:19:11.468534Z","shell.execute_reply.started":"2022-02-24T22:19:09.410912Z","shell.execute_reply":"2022-02-24T22:19:11.465095Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Figure 2. Frequency of oligomer 'A0T0G5C5' in bacterial samples (n = 5.7e7), ± 1 standard error. Length of bars indicates oligomer frequency of occurance in bacterial species, and color indicates bacterial species.","metadata":{}},{"cell_type":"code","source":"# A5T3G0C2 bacterial species composition\nsns.set_context(\"notebook\", font_scale=1.5)\nsns.set_style(\"white\")\nsns.set_color_codes(palette='deep')\nf, ax = plt.subplots(figsize=(8, 6))\nsns.barplot(x = \"A5T3G0C2\", y = \"target\", palette= 'plasma', data = train_dropped)\nplt.tick_params(axis='x', labelsize=15, rotation=-45)\nplt.tick_params(axis='y', labelsize=15, rotation=0)\nplt.xlabel(\"Oligomer Frequency\")\nplt.ylabel(\"Bacterial Species\")","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:19:11.470248Z","iopub.execute_input":"2022-02-24T22:19:11.470616Z","iopub.status.idle":"2022-02-24T22:19:13.45918Z","shell.execute_reply.started":"2022-02-24T22:19:11.470574Z","shell.execute_reply":"2022-02-24T22:19:13.457946Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Figure 3. Frequency of oligomer 'A5T3G0C2' in bacterial samples (n = 5.7e7), ± 1 standard error. Length of bars indicates oligomer frequency of occurance in bacterial species, and color indicates bacterial species.","metadata":{}},{"cell_type":"code","source":"# A3T4G3C0 bacterial species composition\nsns.set_context(\"notebook\", font_scale=1.5)\nsns.set_style(\"white\")\nsns.set_color_codes(palette='deep')\nf, ax = plt.subplots(figsize=(8, 6))\nsns.barplot(x = \"A3T4G3C0\", y = \"target\", palette= 'plasma', data = train_dropped)\nplt.tick_params(axis='x', labelsize=15, rotation=-45)\nplt.tick_params(axis='y', labelsize=15, rotation=0)\nplt.xlabel(\"Oligomer Frequency\")\nplt.ylabel(\"Bacterial Species\")","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:19:13.461188Z","iopub.execute_input":"2022-02-24T22:19:13.461772Z","iopub.status.idle":"2022-02-24T22:19:15.424014Z","shell.execute_reply.started":"2022-02-24T22:19:13.461717Z","shell.execute_reply":"2022-02-24T22:19:15.423009Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Figure 4. Frequency of oligomer 'A3T4G3C0' in bacterial samples (n = 5.7e7), ± 1 standard error. Length of bars indicates oligomer frequency of occurance in bacterial species, and color indicates bacterial species.","metadata":{}},{"cell_type":"code","source":"# A5T4G0C1 bacterial species composition\nsns.set_context(\"notebook\", font_scale=1.5)\nsns.set_style(\"white\")\nsns.set_color_codes(palette='deep')\nf, ax = plt.subplots(figsize=(8, 6))\nsns.barplot(x = \"A5T4G0C1\", y = \"target\", palette= 'plasma', data = train_dropped)\nplt.tick_params(axis='x', labelsize=15, rotation=-45)\nplt.tick_params(axis='y', labelsize=15, rotation=0)\nplt.xlabel(\"Oligomer Frequency\")\nplt.ylabel(\"Bacterial Species\")","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:19:15.425547Z","iopub.execute_input":"2022-02-24T22:19:15.425851Z","iopub.status.idle":"2022-02-24T22:19:17.440939Z","shell.execute_reply.started":"2022-02-24T22:19:15.425817Z","shell.execute_reply":"2022-02-24T22:19:17.439856Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Figure 5. Frequency of oligomer 'A5T4G0C1' in bacterial samples (n = 5.7e7), ± 1 standard error. Length of bars indicates oligomer frequency of occurance in bacterial species, and color indicates bacterial species.","metadata":{}},{"cell_type":"markdown","source":"#  <span style=\"color:#004c6d\">  Stage 6: Machine learning modeling  </span>\n####  “All models are wrong, but some are useful” -George E. P. Box.\n### <span style=\"color:#004c6d\">  6.1 Set dataset features as factors  </span> ","metadata":{}},{"cell_type":"code","source":"# setting sale price as the index column\ntrain_labels = train_dropped['target'].reset_index(drop=True)\n# drop sales price variable; axis=1 drops column not row\ntrain_features = train_dropped.drop(['target'], axis=1)\ntest_features = test\n\n# combine train and test datasets\nall_features = pd.concat([train_features, test_features]).reset_index(drop=True)\n# describe\nall_features.shape\n\n# get numeric data\n# set the numeric datatypes\nnumeric_dtypes = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']\n# save potential values (i) in numeric dictionary if dtype is numeric\nnumeric = []\nfor i in all_features.columns:\n    if all_features[i].dtype in numeric_dtypes:\n        numeric.append(i)\n        \n# check the skew of all numerical features with skew function\n# lower skew is better. If we get high scores, we can transform data to make the data distribution more normal\nskew_features = all_features[numeric].apply(lambda x: skew(x)).sort_values(ascending=False)\n\n# set values with a high skew as being above > 0.5, create dictionary\nhigh_skew = skew_features[skew_features > 0.5]\nskew_index = high_skew.index\n\n# count how many variables have high skew\nprint(\"There are {} variables with a skew > 0.5 :\".format(high_skew.shape[0]))","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:19:17.442502Z","iopub.execute_input":"2022-02-24T22:19:17.442792Z","iopub.status.idle":"2022-02-24T22:19:19.251911Z","shell.execute_reply.started":"2022-02-24T22:19:17.442759Z","shell.execute_reply":"2022-02-24T22:19:19.250989Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### <span style=\"color:#004c6d\">  6.2 Skewed Variables  </span> ","metadata":{}},{"cell_type":"code","source":"# set dataframe for values with high skew\nskewness = pd.DataFrame({'Skew' :high_skew})\n# show values\nskew_features.head(30)","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:19:19.253212Z","iopub.execute_input":"2022-02-24T22:19:19.253442Z","iopub.status.idle":"2022-02-24T22:19:19.264425Z","shell.execute_reply.started":"2022-02-24T22:19:19.253414Z","shell.execute_reply":"2022-02-24T22:19:19.263709Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"fig = plt.figure()\nres = stats.probplot(all_features['A1T0G0C9'], plot=plt)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:19:19.265559Z","iopub.execute_input":"2022-02-24T22:19:19.265794Z","iopub.status.idle":"2022-02-24T22:19:19.962357Z","shell.execute_reply.started":"2022-02-24T22:19:19.265766Z","shell.execute_reply":"2022-02-24T22:19:19.961593Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Figure 6. Ordered values compared with theoretical quantiles of oligomer 'A1T0G0C9'.","metadata":{}},{"cell_type":"code","source":"fig = plt.figure()\nres = stats.probplot(all_features['A0T10G0C0'], plot=plt)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:19:19.96371Z","iopub.execute_input":"2022-02-24T22:19:19.964475Z","iopub.status.idle":"2022-02-24T22:19:20.656857Z","shell.execute_reply.started":"2022-02-24T22:19:19.964431Z","shell.execute_reply":"2022-02-24T22:19:20.655794Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Figure 7. Ordered values compared with theoretical quantiles of oligomer 'A0T10G0C0'.","metadata":{}},{"cell_type":"code","source":"fig = plt.figure()\nres = stats.probplot(all_features['A0T0G9C1'], plot=plt)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:19:20.658329Z","iopub.execute_input":"2022-02-24T22:19:20.65862Z","iopub.status.idle":"2022-02-24T22:19:21.343987Z","shell.execute_reply.started":"2022-02-24T22:19:20.658584Z","shell.execute_reply":"2022-02-24T22:19:21.343007Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Figure 8. Ordered values compared with theoretical quantiles of oligomer 'A0T0G9C1'.","metadata":{}},{"cell_type":"code","source":"skew_cols = train_dropped.select_dtypes(exclude='object').skew().sort_values(ascending=False)\nskew_cols = pd.DataFrame(skew_features.loc[skew_features > 0.75]).rename(columns={0:'Skew before'})\n\n# Box-cox transformation\nt=train_dropped.copy()\nfor i in skew_cols.index.tolist():\n    t[i] = boxcox1p(t[i], boxcox_normmax(t[i] + 1))\n    \nskew_df=pd.concat([skew_cols, t[skew_cols.index].skew()], axis=1).rename(columns={0:'After'})\nskew_df.head(30)","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:19:21.34579Z","iopub.execute_input":"2022-02-24T22:19:21.346349Z","iopub.status.idle":"2022-02-24T22:21:46.033969Z","shell.execute_reply.started":"2022-02-24T22:19:21.346294Z","shell.execute_reply":"2022-02-24T22:21:46.033159Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"fig = plt.figure()\nres = stats.probplot(t['A1T0G0C9'], plot=plt)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:21:46.03562Z","iopub.execute_input":"2022-02-24T22:21:46.036059Z","iopub.status.idle":"2022-02-24T22:21:46.508692Z","shell.execute_reply.started":"2022-02-24T22:21:46.036018Z","shell.execute_reply":"2022-02-24T22:21:46.508047Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Figure 9. Post boxcox transformation ordered values compared with theoretical quantiles of oligomer 'A1T0G0C9'.","metadata":{}},{"cell_type":"code","source":"fig = plt.figure()\nres = stats.probplot(t['A0T10G0C0'], plot=plt)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:21:46.509852Z","iopub.execute_input":"2022-02-24T22:21:46.510824Z","iopub.status.idle":"2022-02-24T22:21:46.980314Z","shell.execute_reply.started":"2022-02-24T22:21:46.51078Z","shell.execute_reply":"2022-02-24T22:21:46.979195Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Figure 10. Post boxcox transformation ordered values compared with theoretical quantiles of oligomer 'A0T10G0C0'.","metadata":{}},{"cell_type":"code","source":"fig = plt.figure()\nres = stats.probplot(t['A0T0G9C1'], plot=plt)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:21:46.981796Z","iopub.execute_input":"2022-02-24T22:21:46.982067Z","iopub.status.idle":"2022-02-24T22:21:47.457866Z","shell.execute_reply.started":"2022-02-24T22:21:46.982036Z","shell.execute_reply":"2022-02-24T22:21:47.456624Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Figure 11. Post boxcox transformation ordered values compared with theoretical quantiles of oligomer 'A0T0G9C1'.","metadata":{}},{"cell_type":"markdown","source":"### <span style=\"color:#004c6d\">  6.3 Standardize Data </span>","metadata":{}},{"cell_type":"code","source":"X=t.drop('target', axis=1)\nX_scaled=pd.DataFrame(StandardScaler().fit_transform(X), columns=X.columns)\nX_scaled.head()","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:21:47.459693Z","iopub.execute_input":"2022-02-24T22:21:47.460064Z","iopub.status.idle":"2022-02-24T22:21:49.017438Z","shell.execute_reply.started":"2022-02-24T22:21:47.460008Z","shell.execute_reply":"2022-02-24T22:21:49.016586Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### <span style=\"color:#004c6d\">  6.4 Principal Components Analysis </span>\n\nNow that my data has been converted into factors, I have processed data for missing and duplicate values, and standardized data asymmetry, I will be able to create a model for my study objective. To classify the 10 bacterial species from the test dataset, I will use principal components analysis (PCA). Principal component analysis (PCA) is the process of creating new varibles called Principal Components from a linear combination of metrics (oligomer sequences) that capture the greatest amount of variance within the dataset. Each Principal Component successively finds the maximum variance in the data projected along the orthogonal axis of the previous component, creating a new set of uncorrelated features that collectively explain the variability within the data. PCA allows us to summarize information by smaller “summary indices”, which are more easily visualized and analyzed.  ","metadata":{}},{"cell_type":"code","source":"# set cumulative components\npca = PCA(n_components=286).fit(X_scaled)\n# set multiplier\npca_sum=pd.Series(np.cumsum(pca.explained_variance_ratio_)).mul(100)\n# set index and dataframe\npca_sum.index = np.arange(1, len(pca_sum)+1)\n# set individual componenents\nind=pd.Series(pca.explained_variance_ratio_).mul(100)\n# set index and dataframe\nind.index = np.arange(1, len(ind)+1)\n\n# PCA graph\nfig = go.Figure(\n    layout=go.Layout(\n        updatemenus=[dict(type=\"buttons\", direction=\"left\", x=0.15, y=1.2, showactive=False)],\n        xaxis=dict(range=[1, 287],\n                   autorange=False, tickwidth=2),\n        yaxis=dict(range=[0, 100],\n                   autorange=False)))\n\nfig.add_trace(go.Scatter(x=ind.index[:1], y=ind[:1], line=dict(color='#5f007f', width=3), \n                         visible=True, fill='tozeroy', opacity=0.8,\n                         hovertemplate = 'Variance Explained = %{y:.2f}%<br>Principal Component %{x:.0f}',\n                         name='Individual'))\nfig.add_trace(go.Scatter(x=pca_sum.index[:1], y=pca_sum[:1], line=dict(color='#ADD8E6', width=3), \n                         visible=True, fill='tonexty', opacity=0.7,\n                         hovertemplate = 'Variance Explained = %{y:.1f}%<br>Number of Principal Components = %{x:.0f}',\n                         name='Cumulative'))\n\nfig.update(frames=[go.Frame(data=[\n    go.Scatter(x=ind.index[:i], y=ind[:i]),\n    go.Scatter(x=pca_sum.index[:i], y=pca_sum[:i])])\n                   for i in range(1, 287)])\n\nfig.update_yaxes(title = 'Variance Explained', showline=True, ticksuffix='%', range=[0,105])\n\nfig.update_layout(title='Principal Components Explained Variance', \n                  xaxis_title=\"Number of Principal Components\",\n                  hovermode=\"x unified\", width=700,\n                  legend=dict(orientation=\"v\", yanchor=\"bottom\", y=1.08, xanchor=\"right\", x=.99, title=\"\"),\n                  updatemenus=[dict(buttons=list(\n                      [dict(label=\"Play\", method=\"animate\", \n                            args=[None, {\"frame\": {\"duration\":15, \"redraw\": False}},{\"fromcurrent\": True}]),\n                       dict(label=\"Pause\", method=\"animate\", \n                            args=[{\"frame\": {\"duration\": 0, \"redraw\": False}},{\"mode\": \"immediate\"},\n                                  {\"transition\": {\"duration\": 0}}])]))])\nfig.show()","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:21:49.019243Z","iopub.execute_input":"2022-02-24T22:21:49.019751Z","iopub.status.idle":"2022-02-24T22:21:54.946349Z","shell.execute_reply.started":"2022-02-24T22:21:49.019706Z","shell.execute_reply":"2022-02-24T22:21:54.945063Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Figure 12. The variance captured from calculated Principal Components, both cumulatively and individually, indicated by the blue and purple lines. The cumulative PC score reaches an infection point early in the number of Principal Components.","metadata":{}},{"cell_type":"code","source":"# Variable significance: important oligomer sequences\ndf=pd.DataFrame(abs(pca.components_.T), columns=['PC'+str(i+1) for i in range(286)], index=X_scaled.columns)\npca_ind=pd.Series(pca.explained_variance_ratio_)\nvar_pca=[]\nfor i,j in zip(df.columns.tolist(), pca_ind):\n    k=df[i].nlargest(1)*j\n    var_pca.append(pd.DataFrame({'Principal Component':str(i[2:]),'Gene':k.index,'Var':k[0]}))\nvar_pca=pd.concat(var_pca).reset_index(drop=True)\nplot_df=var_pca.iloc[:10,:]\n\npal = sns.color_palette(\"plasma\", 14).as_hex()[1:11]\nfig = px.bar(plot_df, x='Gene', y='Var', text='Var', color='Principal Component', \n             color_discrete_sequence=pal, opacity=0.7)\nfig.update_traces(texttemplate='%{text:,.3f}', textposition='outside',\n                  marker_line=dict(width=1, color='#28221D'))\nfig.show()","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:21:54.948096Z","iopub.execute_input":"2022-02-24T22:21:54.948398Z","iopub.status.idle":"2022-02-24T22:21:56.194652Z","shell.execute_reply.started":"2022-02-24T22:21:54.948361Z","shell.execute_reply":"2022-02-24T22:21:56.193747Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Figure 13. Composition of oligomer sequency in Principal Components of bacterial data analysis. Height of bar indicates the relative makeup of the Principal Component by the genetic sequence, and color indicates different oligomer sequence.","metadata":{}},{"cell_type":"code","source":"# PC 1 and 2\npca = PCA(n_components=10).fit_transform(X_scaled)\npca_df=pd.DataFrame(data=pca, columns=['PC'+str(i+1) for i in range(0,10)]).reset_index(drop=True)\nspecies=train_dropped.target.reset_index(drop=True).str.replace('_', ' ') \npca_df=pd.concat([species, pca_df], axis=1)\npca_df['map'] = pca_df['target'].map(pca_df['target'].value_counts())\npca_df = pca_df.sort_values(by='map', ascending=False).drop('map', axis=1)\n\npal = sns.color_palette(\"plasma\", 12).as_hex()[:10]\nfig = px.scatter(pca_df, x='PC1', y='PC2', color='target', color_discrete_sequence=pal, opacity=0.4)\nfig.update_traces(marker_size=7,\n                  hovertemplate=\"Principal Component 1 = %{x}<br>Principal Component 2 = %{y}\")\nfig.update_layout(title='Bacteria Species Projected onto Components 1 and 2', legend_title='', \n                  xaxis_title='Component 1 (variance explained = 31.9%)', \n                  yaxis_title='Component 2 (variance explained = 20.4%)',\n                  width=700, height=600)\nfig.show()","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:21:56.196032Z","iopub.execute_input":"2022-02-24T22:21:56.196366Z","iopub.status.idle":"2022-02-24T22:22:01.955437Z","shell.execute_reply.started":"2022-02-24T22:21:56.19633Z","shell.execute_reply":"2022-02-24T22:22:01.952971Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Figure 14. Principal components 1 and 2 of oligomer bacterial samples. Bacterial species are indicated by color.","metadata":{}},{"cell_type":"markdown","source":"To interprete this figure, we can see the bacterial oligomer score vectors graphed on 2 Principal Component axes. The separation of species is greatest for species *Klebsiella pneumoniae* and *Bacteroides fragilis*, with varying separatation across species by commonly shared oligomer sequences. Variance in data is explained by 31.9% and 20.4% in Principal Components 1 and 2. Below are the projections onto the first three principal components for each species.","metadata":{}},{"cell_type":"code","source":"# PC 1-3\ns=pca_df.target.unique()\nrgb=[]\nfor i in pal:\n    rgb.append('rgb' + str(mpl.to_rgb(i)))\n\nfig = make_subplots(rows=5, cols=2,\n                    specs=[[{'type': 'scatter3d'}, {'type': 'scatter3d'}],\n                           [{'type': 'scatter3d'}, {'type': 'scatter3d'}], \n                           [{'type': 'scatter3d'}, {'type': 'scatter3d'}],\n                           [{'type': 'scatter3d'}, {'type': 'scatter3d'}],\n                           [{'type': 'scatter3d'}, {'type': 'scatter3d'}]],\n                    horizontal_spacing = 0.1, vertical_spacing = 0.05,\n                    subplot_titles=(s[0],s[1],s[2],s[3],s[4],\n                                    s[5],s[6],s[7],s[8],s[9]))\n\np1=pca_df[pca_df.target=='Bacteroides fragilis']\nfig.add_trace(go.Scatter3d(x=p1.PC1, y=p1.PC2, z=p1.PC3, mode='markers', showlegend=False,\n                           marker=dict(size=3, color=pal[0], opacity=0.4, line_width=1, line_color=rgb[0])),\n              row=1, col=1)\np2=pca_df[pca_df.target=='Campylobacter jejuni']\nfig.add_trace(go.Scatter3d(x=p2.PC1, y=p2.PC2, z=p2.PC3, mode='markers', showlegend=False,\n                           marker=dict(size=3, color=pal[1], opacity=0.3, line_width=1, line_color=rgb[1])),\n              row=1, col=2)\np3=pca_df[pca_df.target=='Klebsiella pneumoniae']\nfig.add_trace(go.Scatter3d(x=p3.PC1, y=p3.PC2, z=p3.PC3, mode='markers', showlegend=False,\n                           marker=dict(size=3, color=pal[2], opacity=0.3, line_width=1, line_color=rgb[2])),\n              row=2, col=1)\np4=pca_df[pca_df.target=='Streptococcus pneumoniae']\nfig.add_trace(go.Scatter3d(x=p4.PC1, y=p4.PC2, z=p4.PC3, mode='markers', showlegend=False,\n                           marker=dict(size=3, color=pal[3], opacity=0.3, line_width=1, line_color=rgb[3])),\n              row=2, col=2)\np5=pca_df[pca_df.target=='Staphylococcus aureus']\nfig.add_trace(go.Scatter3d(x=p5.PC1, y=p5.PC2, z=p5.PC3, mode='markers', showlegend=False,\n                           marker=dict(size=3, color=pal[4], opacity=0.3, line_width=1, line_color=rgb[4])),\n              row=3, col=1)\np6=pca_df[pca_df.target=='Streptococcus pyogenes']\nfig.add_trace(go.Scatter3d(x=p6.PC1, y=p6.PC2, z=p6.PC3, mode='markers', showlegend=False,\n                           marker=dict(size=3, color=pal[5], opacity=0.35, line_width=1, line_color=rgb[5])),\n              row=3, col=2)\np7=pca_df[pca_df.target=='Salmonella enterica']\nfig.add_trace(go.Scatter3d(x=p7.PC1, y=p7.PC2, z=p7.PC3, mode='markers', showlegend=False,\n                           marker=dict(size=3, color=pal[6], opacity=0.35, line_width=1, line_color=rgb[6])),\n              row=4, col=1)\np8=pca_df[pca_df.target=='Enterococcus hirae']\nfig.add_trace(go.Scatter3d(x=p8.PC1, y=p8.PC2, z=p8.PC3, mode='markers', showlegend=False,\n                           marker=dict(size=3, color=pal[7], opacity=0.35, line_width=1, line_color=rgb[7])),\n              row=4, col=2)\np9=pca_df[pca_df.target=='Escherichia coli']\nfig.add_trace(go.Scatter3d(x=p9.PC1, y=p9.PC2, z=p9.PC3, mode='markers', showlegend=False,\n                           marker=dict(size=3, color=pal[8], opacity=0.35, line_width=1, line_color=rgb[8])),\n              row=5, col=1)\np10=pca_df[pca_df.target=='Escherichia fergusonii']\nfig.add_trace(go.Scatter3d(x=p10.PC1, y=p10.PC2, z=p10.PC3, mode='markers', showlegend=False,\n                           marker=dict(size=3, color=pal[9], opacity=0.35, line_width=1, line_color=rgb[9])),\n              row=5, col=2)\nfig.update_traces(hovertemplate='Component 1 = %{x}<br>Component 2 = %{y}<br>Component 3 = %{z}<extra></extra>')\nfig.update_layout(title='Bacteria Species Projected onto the first 3 Principal Components',\n                  scene1=dict(aspectmode='cube',xaxis_title='PC 1',yaxis_title='PC 2',zaxis_title='PC 3'), \n                  scene2=dict(aspectmode='cube', xaxis_title='PC 1',yaxis_title='PC 2',zaxis_title='PC 3'), \n                  scene3=dict(aspectmode='cube', xaxis_title='PC 1',yaxis_title='PC 2',zaxis_title='PC 3'), \n                  scene4=dict(aspectmode='cube', xaxis_title='PC 1',yaxis_title='PC 2',zaxis_title='PC 3'),\n                  scene5=dict(aspectmode='cube', xaxis_title='PC 1',yaxis_title='PC 2',zaxis_title='PC 3'), \n                  scene6=dict(aspectmode='cube', xaxis_title='PC 1',yaxis_title='PC 2',zaxis_title='PC 3'),\n                  scene7=dict(aspectmode='cube', xaxis_title='PC 1',yaxis_title='PC 2',zaxis_title='PC 3'),\n                  scene8=dict(aspectmode='cube', xaxis_title='PC 1',yaxis_title='PC 2',zaxis_title='PC 3'),\n                  scene9=dict(aspectmode='cube', xaxis_title='PC 1',yaxis_title='PC 2',zaxis_title='PC 3'), \n                  scene10=dict(aspectmode='cube', xaxis_title='PC 1',yaxis_title='PC 2',zaxis_title='PC 3'),\n                  height=2500, width=800)\nfig.show()","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:22:01.957652Z","iopub.execute_input":"2022-02-24T22:22:01.958683Z","iopub.status.idle":"2022-02-24T22:22:02.746396Z","shell.execute_reply.started":"2022-02-24T22:22:01.958619Z","shell.execute_reply":"2022-02-24T22:22:02.745004Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Figure 15. Principal Components 1, 2, and 3 of oligomer frequency in bacterial species. Color indicated bacterial species, and spread of data indicates a simplified summary of oligomer characteristics per species.","metadata":{}},{"cell_type":"markdown","source":"### <span style=\"color:#004c6d\">  6.5 Model Prediction </span>\n\nTo predict the species of bacteria, I will use two types of models, one with the full set of features in the dataset and one with a subset of features using the first 100 Principal Components. These components cumulatively explain over 85% of the variation in the data. ","metadata":{}},{"cell_type":"code","source":"s = StandardScaler()\nenc=LabelEncoder()\ny=train_dropped.target\ny=enc.fit_transform(y)\ntrain_dropped.drop('target', axis=1, inplace=True)\n\nX_train, X_val, y_train, y_val = train_test_split(train_dropped, y, test_size=0.2, shuffle=True, \n                                                  stratify=y, random_state=21)\n\nX_train_scaled = s.fit_transform(X_train)\nX_val_scaled = s.transform(X_val)\nX_test_scaled = s.transform(test)\n\npca = PCA(n_components=100)\nX_train_pca=pca.fit_transform(X_train_scaled)\nX_val_pca=pca.transform(X_val_scaled)\nX_test_pca=pca.transform(X_test_scaled)\n\nprint(\"Train Shape: {} {}\".format(X_train_pca.shape, y_train.shape))\nprint(\"Validation Shape: {} {}\".format(X_val_pca.shape, y_val.shape))\nprint(\"Test Shape: {}\\n\".format(X_test_pca.shape))\n\net_pca=ExtraTreesClassifier(n_estimators=500,\n                            class_weight='balanced',\n                            random_state=92).fit(X_train_pca, y_train)\nprint(et_pca)\n\ny_preds=et_pca.predict(X_val_pca)\ny_probs=et_pca.predict_proba(X_val_pca)\nval_acc=accuracy_score(y_true=y_val, y_pred=y_preds)\nval_auc=roc_auc_score(y_true=y_val, y_score=y_probs, average='weighted', multi_class='ovr')\nc=classification_report(y_val, y_preds, target_names=enc.classes_, output_dict=True)\nc=pd.DataFrame(c).T.iloc[:10,][['f1-score', 'precision', 'recall', 'support']]\nval_f1=c['f1-score'].mean()\n\nprint('\\nModel Accuracy = {:.2f}%\\nF1-Score = {:.2f}%\\nArea Under the Curve = {:.3f}\\n'\\\n      .format(val_acc*100, val_f1*100, val_auc))\n\nc[['f1-score', 'precision', 'recall']]=c[['f1-score', 'precision', 'recall']].mul(100)\nc.sort_values('f1-score', ascending=False).style\\\n.background_gradient(cmap='flare_r', subset=['f1-score'])\\\n.format({'f1-score':'{:,.1f}%', 'precision':'{:,.1f}%', 'recall':'{:,.1f}%', \"support\": \"{:,.0f}\"})","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:22:02.752584Z","iopub.execute_input":"2022-02-24T22:22:02.753717Z","iopub.status.idle":"2022-02-24T22:24:50.031147Z","shell.execute_reply.started":"2022-02-24T22:22:02.753656Z","shell.execute_reply":"2022-02-24T22:24:50.029969Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"fpr = {}\ntpr = {}\nroc_auc = {}\nthresh = {}\n\nspecies=c.sort_values('f1-score', ascending=False).index.str.replace('_', ' ')\nfor i in range(len(species)):    \n    fpr[i], tpr[i], thresh[i] = roc_curve(y_val, y_probs[:,i], pos_label=i)\n    roc_auc[i] = auc(fpr[i], tpr[i])\n\nfig = go.Figure()\nfor i,j in zip(enumerate(species), pal):\n    fig.add_trace(go.Scatter(x=fpr[i[0]], y=tpr[i[0]], line=dict(color=j, width=3), opacity=0.7,\n                             hovertemplate = 'True positive rate = %{y:.3f}, False positive rate = %{x:.3f}',\n                             name='{} AUC = {:.3f}'.format(i[1],roc_auc[i[0]])))\nfig.add_shape(type=\"line\", xref=\"x\", yref=\"y\", x0=0, y0=0, x1=1, y1=1, \n              line=dict(color=\"Black\", width=1, dash=\"dot\"))\nfig.update_layout(title=\"Multiclass ROC Curves<br>of Bacteria Species\", hovermode=\"x unified\", \n                  hoverlabel = dict(bgcolor=\"white\",font_size=12), xaxis=dict(zeroline=False, hoverformat=\".2f\"),\n                  xaxis_title='False Positive Rate (1 - Specificity)', yaxis_title='True Positive Rate (Sensitivity)',\n                  legend=dict(y=.1, x=.98, xanchor=\"right\",bordercolor=\"black\", borderwidth=.5, font=dict(size=12)),\n                  height=550, width=700)\nfig.show()","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:24:50.033067Z","iopub.execute_input":"2022-02-24T22:24:50.033304Z","iopub.status.idle":"2022-02-24T22:24:50.14995Z","shell.execute_reply.started":"2022-02-24T22:24:50.033277Z","shell.execute_reply":"2022-02-24T22:24:50.149061Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Figure 16. ROC curve of PCA model of bacterial species analysis.","metadata":{}},{"cell_type":"code","source":"test_preds=et_pca.predict(X_test_pca)\ntarget=enc.inverse_transform(test_preds)\nsub_pca=pd.DataFrame({'row_id':[i for i in range(int(2e5),int(3e5))], 'target':target})\nbact=sub_pca.target.value_counts(normalize=True).reset_index()\nbact.target=bact.target.mul(100).sort_values(ascending=False)","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:24:50.151188Z","iopub.execute_input":"2022-02-24T22:24:50.151426Z","iopub.status.idle":"2022-02-24T22:25:07.666241Z","shell.execute_reply.started":"2022-02-24T22:24:50.151397Z","shell.execute_reply":"2022-02-24T22:25:07.665178Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"sub_pca.to_csv(\"submission_pca.csv\", index=False)","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:25:07.667804Z","iopub.execute_input":"2022-02-24T22:25:07.668478Z","iopub.status.idle":"2022-02-24T22:25:07.930651Z","shell.execute_reply.started":"2022-02-24T22:25:07.66843Z","shell.execute_reply":"2022-02-24T22:25:07.929598Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### <span style=\"color:#004c6d\">  6.6 Model Application </span>","metadata":{}},{"cell_type":"code","source":"print(\"Train Shape: {} {}\".format(X_train_scaled.shape, y_train.shape))\nprint(\"Validation Shape: {} {}\".format(X_val_scaled.shape, y_val.shape))\nprint(\"Test Shape: {}\\n\".format(X_test_scaled.shape))\n\net_all=ExtraTreesClassifier(n_estimators=500,  \n                            class_weight='balanced', \n                            random_state=21).fit(X_train_scaled, y_train)\nprint(et_all)\n\ny_preds=et_all.predict(X_val_scaled)\ny_probs=et_all.predict_proba(X_val_scaled)\nval_acc=accuracy_score(y_true=y_val, y_pred=y_preds)\nval_auc=roc_auc_score(y_true=y_val, y_score=y_probs, average='weighted', multi_class='ovr')\nc=classification_report(y_val, y_preds, target_names=enc.classes_, output_dict=True)\nc=pd.DataFrame(c).T.iloc[:10,][['f1-score', 'precision', 'recall', 'support']]\nval_f1=c['f1-score'].mean()\n\nprint('Model Accuracy = {:.2f}%\\nF1-Score = {:.2f}%\\nArea Under the Curve = {:.3f}\\n'\\\n      .format(val_acc*100, val_f1*100, val_auc))\n\nc[['f1-score', 'precision', 'recall']]=c[['f1-score', 'precision', 'recall']].mul(100)\nc.sort_values('f1-score', ascending=False).style\\\n.background_gradient(subset=['f1-score'])\\\n.format({'f1-score':'{:,.1f}%', 'precision':'{:,.1f}%', 'recall':'{:,.1f}%', \"support\": \"{:,.0f}\"})","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:25:07.932258Z","iopub.execute_input":"2022-02-24T22:25:07.932926Z","iopub.status.idle":"2022-02-24T22:29:09.616526Z","shell.execute_reply.started":"2022-02-24T22:25:07.93287Z","shell.execute_reply":"2022-02-24T22:29:09.615617Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Now submit to the competition!","metadata":{}},{"cell_type":"code","source":"test_preds=et_all.predict(X_test_scaled)\ntarget=enc.inverse_transform(test_preds)\nsub=pd.DataFrame({'row_id':[i for i in range(int(2e5),int(3e5))], 'target':target})\nbact=sub.target.value_counts(normalize=True).reset_index()\nbact.target=bact.target.mul(100).sort_values(ascending=False)","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:29:09.618317Z","iopub.execute_input":"2022-02-24T22:29:09.618779Z","iopub.status.idle":"2022-02-24T22:29:40.187448Z","shell.execute_reply.started":"2022-02-24T22:29:09.61874Z","shell.execute_reply":"2022-02-24T22:29:40.18669Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"sub.to_csv(\"submission.csv\", index=False)","metadata":{"execution":{"iopub.status.busy":"2022-02-24T22:29:40.188686Z","iopub.execute_input":"2022-02-24T22:29:40.18931Z","iopub.status.idle":"2022-02-24T22:29:40.443534Z","shell.execute_reply.started":"2022-02-24T22:29:40.189267Z","shell.execute_reply":"2022-02-24T22:29:40.442815Z"},"trusted":true},"execution_count":null,"outputs":[]}]}