{"cells":[{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true},"cell_type":"code","source":"# imports\nimport pandas as pd\nimport numpy as np\nimport plotly.express as px\nimport plotly\nimport matplotlib.pyplot as plt\nimport seaborn as sns\nimport plotly.graph_objects as go\nplotly.offline.init_notebook_mode (connected = True)","execution_count":null,"outputs":[]},{"metadata":{"_uuid":"d629ff2d2480ee46fbb7e2d37f6b5fab8052498a","_cell_guid":"79c7e3d0-c299-4dcb-8224-4455121ee9b0","trusted":true},"cell_type":"code","source":"# reading the features data\ntrain_data = pd.read_csv('../input/lish-moa/train_features.csv')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## High Level Exploration","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"train_data.head(10)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"train_data.shape","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"train_data.info()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"\n- There are a total of 876 columns and 23814 rows.<br>\n- 872 float columns, 1 integer and 3 object columns.<br>\n- The object columns are: sig_id, cp_time, cp_dose.<br>","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"g = (train_data.columns).map(lambda x: x.startswith('g-')).tolist()\nc = (train_data.columns).map(lambda x: x.startswith('c-')).tolist()\n\nprint(f'Number of gene expression variables {sum(g)},\\nNumber of cell viability variables {sum(c)}')\n","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"train_data.describe()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"# categorical variables (except sig_id)\ntrain_data.columns[train_data.nunique() < 10]","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"\n- cp_type, cp_time and cp_dose are categorical in nature.","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"# changing to object\ntrain_data['cp_time'] = train_data['cp_time'].astype(object)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## Categorical Features\n\n- cp_time: It is the treatment duration, let's analyze it further.<br>\n- cp_type: It indicates if a sample was treated with a compound or a control perturbation<br>\n- cp_dose: If the dise was high or low     ","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"fig, axes = plt.subplots(figsize = (20,4), nrows = 1, ncols = 3)\nsns.countplot(train_data['cp_time'], ax = axes[0])\nsns.countplot(train_data['cp_type'], ax = axes[1])\nsns.countplot(train_data['cp_dose'], ax = axes[2])\n\ntitles = ['CP TIME COUNTS', 'CP TYPE COUNTS', 'CP DOSE COUNTS']\n\nfor a in range(3):\n    axes[a].set_title(titles[a])\n\nplt.show()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"\n- cp_time has almost a balanced distribution, 48 hour treatments seem to be most frequent with ~ 8000 counts.<br>\n- cp_type has an imbalanced distribution, most sample are treated with trt_cp.<br>\n- cp_dose also has similar distribution in the classes, D1 dose is given to most of the sample.    ","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"## Continuous Variables<br>\n\n\n- Gene Expression: g-0 to g-771    <br>\n- Cell Viability: c-0 to c-99    ","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"def plot_density(variable = 'g-', rows = 3, cols = 3):\n    \"\"\"\n    Randomlt selects some columns for the given variable and plots their density function.\n    \"\"\"\n    \n    fig,axes = plt.subplots(figsize = (15,12), nrows = rows, ncols = cols)\n    \n    if variable == 'g-':\n        m = 771\n        color  = 'coral'\n    else:\n        m = 99\n        color = 'green'\n    \n    v = np.random.randint(0, m, rows*cols)\n    \n    k = 0\n    for i in range(rows):\n        for j in range(cols):\n            sns.kdeplot(train_data[variable + str(v[k])], ax = axes[i][j], shade = True, color = color)\n            axes[i][j].set_title(f'{variable + str(v[k])} Distribution')\n            k+=1\n    plt.show()\n    ","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### <b> Gene Expression","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"plot_density('g-', rows = 3, cols = 3)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"\n- All the variables are centred around 0 are close to normal distribution. <br>\n- A slight skew can be detected in the variables, g-61, g-369, g-444 etc are left skewed whereas g-289, g-264 etc are right skewed<br>","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"## Cell Viability","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"plot_density('c-')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"\n- Here as well the variables have been centred to 0 and have almost normal distributions.<br>\n- Most of the variables are left skewed and have a peak at -10, this could be an anomaly.","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"## Signal Visualisation<br>\n\nTaking inspiration from this notebook: https://www.kaggle.com/artgor/code-for-live-pair-coding by Andrew Lukyanenko\nI have visualised the gene expressions and cell viabilities for single samples. They look like some sort of signals.","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"def plot_signal(variable = 'g-',  rows = 3, ma1 = 10, ma2 = 30):\n    \"\"\"\n    Plots the variable for randomly selected samples with transformations\n    \"\"\"    \n    \n    if variable == 'g-':\n        m = 771\n        title = 'Gene Expression'\n    else:\n        m = 99\n        title = 'Cell Viability'\n\n    colors = ['#9b59b6', '#00a8ff', '#A3CB38', '#B53471']    \n    v = np.random.randint(0, m, rows)\n    fig,axes = plt.subplots(figsize = (18,12), nrows = rows, ncols = 4)\n    \n    plt.suptitle(title, size = 30)\n    for i in range(len(v)):        \n        train_data.loc[:,train_data.columns.str.startswith(variable)].iloc[v[i]].plot(color = colors[0], \n                                                                                   ax = axes[i][0],title = f'Sample {v[i]} Signals')\n        train_data.loc[:,train_data.columns.str.startswith(variable)].iloc[v[i]].sort_values().plot(color = colors[1], \n                                                                                                 ax = axes[i][1], title = f'Sample {v[i]} Sorted Signals')\n        train_data.loc[:,train_data.columns.str.startswith(variable)].iloc[v[i]].rolling(ma1).mean().plot(color = colors[2], \n                                                                                                      ax = axes[i][2], title = f'Sample {v[i]} Window {ma1} Roling Mean')\n        train_data.loc[:,train_data.columns.str.startswith(variable)].iloc[v[i]].rolling(ma2).mean().plot(color = colors[3], \n                                                                                                     ax = axes[i][3], title = f'Sample {v[i]} Window {ma2} Roling Mean')\n    \n    plt.show()","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"plot_signal('g-')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"\n- The first columns shows the raw gene signals for various samples, it is noisy but there definitely seems to be some sort of a pattern. To me it looks like what we get after doing differrencing on a time series data (or any signal like data to make it stationary).<br>\n- In the 3rd and 4th column I have done a rolling mean with 10 and 30 period windows respectively and we can definitely see some sort of a pattern emerging in each sample. We can use sequence models to capture this data. <br>\n- Another interesting thing that was pointed out by Andrew was that we can sort the gene expressions. On doing so in column 2 we can see that a very clean graph emerges. It looks like a variant of the `tangent function` <br>\n\nUsing these insights we can calculate some features which might help us in modelling.","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"plot_signal('c-')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"- In cell viability as well we can see that sort sort of patterns are visible on doing rolling mean.<br>\n- The sorted signals are less smooth than the gene expressions however they also are following a similar function.","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"### ...to be continued","execution_count":null}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}