{"cells":[{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true},"cell_type":"markdown","source":"# OpenVaccine : Literature Survey"},{"metadata":{},"cell_type":"markdown","source":"This notebook does not contain any code. This is a read through notebook for reference regarding OpenVaccine Challenge. I had compiled them for my perusal. But for the benefit of all competitors, decided to make it public."},{"metadata":{},"cell_type":"markdown","source":"## References:\n\n1. [OpenVaccine Homepage in Kaggle](https://www.kaggle.com/c/stanford-covid-vaccine)\n1. [Discussion Thread by @hengck23](https://www.kaggle.com/c/stanford-covid-vaccine/discussion/182652)\n"},{"metadata":{},"cell_type":"markdown","source":"This notebook contains excerpts from various sources related to OpenVaccine Challenge. As far as possible, an effort is made to avoid excerpts from copyrighted material. However, accidentally, if any copyright violation is noticed anywhere in the notebook, kindly mention in comment so that the same can be removed."},{"metadata":{},"cell_type":"markdown","source":"## [About Eterna project and RNA Vaccines](https://medium.com/eternaproject/how-to-build-a-better-vaccine-from-the-comfort-of-your-own-web-browser-233343e0210d)"},{"metadata":{},"cell_type":"markdown","source":"### How stable can we make a messenger RNA?\n\nWe score molecules based on the Average Unpaired Probability (AUP) over all the bases. The p(unpaired) for each nucleotide is predicted by algorithms designed for this purpose and used widely in RNA structure prediction. A lower AUP means that the molecule is predicted to be more stable."},{"metadata":{},"cell_type":"markdown","source":"## [Competition Sponsor's Paper](https://www.biorxiv.org/content/10.1101/2020.08.22.262931v1)"},{"metadata":{},"cell_type":"markdown","source":"A largely unexplored strategy to reduce mRNA hydrolysis is to redesign RNAs to form double-stranded regions, which are protected from in-line cleavage and enzymatic degradation, while coding for the same proteins. The amount of stabilization that this strategy can deliver and the most effective algorithmic approach to achieve stabilization remain poorly understood. Motivated by the need for stabilized COVID-19 mRNA vaccines, we present simple calculations for estimating RNA stability against hydrolysis, and a model that links the average unpaired probability of an mRNA, or AUP, to its overall rate of hydrolysis."},{"metadata":{},"cell_type":"markdown","source":"The AUP value is a number between 0 and 1 that reflects the overall ‘structuredness’ of the RNA. Lower values correspond to lower probability of being unpaired, and therefore RNA molecules less susceptible to degradation. How to handle AUP for strands with different sequence lengths - Required in this challenge since we need to predict for 130 base given 107 base training data with 68 bases being used. Since this will have an impact on private LB scores, please go through the paper to find out."},{"metadata":{},"cell_type":"markdown","source":"## [Paper on BBPS for deriving Secondary Structure](https://bmcbioinformatics.biomedcentral.com/track/pdf/10.1186/s12859-018-2018-4)"},{"metadata":{},"cell_type":"markdown","source":"Secondary structures of RNA are known to be thermo-dynamically fluctuated and the number of possible sec-ondary structures of RNA are huge. We can predict thesecondary structure by software tools, but on Boltzmanndistribution of the secondary structure the probability ofthe ‘best’ secondary structure predicted is usually verysmall"},{"metadata":{},"cell_type":"markdown","source":"The base-pairing probabilities (BPPs), which are often greater than eighty percent, are convenient to observe the local stability ofthe secondary structures. Furthermore, the estimators based on maximum expected accuracy, such as the MEA(Maximum Expected Accuracy) estimator and the γ-centroid estimator, can be calculated by posteriordecoding of BPPs without using further information of the RNA sequence."},{"metadata":{},"cell_type":"markdown","source":"## [RNAonGraph](https://www.biorxiv.org/content/10.1101/2020.02.11.931030v1)"},{"metadata":{},"cell_type":"markdown","source":"To better model RNA secondary structures and to ease the subsequent representational learning detail, we propose to fill in the usually binary adjacency matrix with base pairing probabilities considering all possible secondary structures from the ensemble of RNA stochastic folding."},{"metadata":{},"cell_type":"markdown","source":"\"Our GNN model takes in as input an adjacency matrix that specifies the base-pairing probabilities, as well as the one-hot encoded nucleotide sequence\""},{"metadata":{},"cell_type":"markdown","source":"## [cDeepbind](https://www.biorxiv.org/content/10.1101/345140v1)"},{"metadata":{},"cell_type":"markdown","source":"We propose two deep architectures, one a lightweight convolutional network for transcriptome wide inference and another a Long Short-Term Memory(LSTM) network that is suitable for small batches of data. We incorporate computationally predicted secondary structure features as input to our models and show its effectiveness in boosting prediction performance."},{"metadata":{},"cell_type":"markdown","source":"The models take as input a sequence s = (s1, …, sn) from the alphabet A = {A, G, C, U} and a structural annotation vector r = (r1, …, rn) where ri ∈ R5 and denotes the probability of the position being in the contexts (paired(P), hairpin loop(H), inner loop(I), multi-loop(M) or external region(E)). "},{"metadata":{},"cell_type":"markdown","source":"The model training pipeline consists of automatic hyperparameter training using cross validation followed training an ensemble of models with the best hyperparameter from the previous step on the entire training set. We use an ensemble of five models as our final predictive model by averaging the predictions across the ensemble. We find that the most important hyperparameters for the model were the learning rate, rate of weight decay, number of LSTM units and variance of the random normal initializer for the weights. We use batch-size scaling which is a new alternative to learning rate decay, to accelerate training. We start training with a minibatch size of 500 and double the batch size after every 5 epochs. We found during cross-validation that loss saturates after 15 epochs and we train all our models for 15 epochs. "},{"metadata":{},"cell_type":"markdown","source":"## [Structure Inference through Chemical Mapping](https://www.pnas.org/content/114/37/9876) -- Sponsor's Paper"},{"metadata":{},"cell_type":"markdown","source":"We describe a multidimensional chemical-mapping method called “mutate-and-map read out through next-generation sequencing” (M2-seq) that takes advantage of sparsely mutated nucleotides to induce structural perturbations at partner nucleotides and then detects these events through dimethyl sulfate (DMS) probing and mutational profiling. In special cases, fortuitous errors introduced during DNA template preparation and RNA transcription are sufficient to give M2-seq helix signatures; these signals were previously overlooked or mistaken for correlated double-DMS events. When mutations are enhanced through error-prone PCR, in vitro M2-seq experimentally resolves 33 of 68 helices in diverse structured RNAs including ribozyme domains, riboswitch aptamers, and viral RNA domains with a single false positive. These inferences do not require energy minimization algorithms and can be made by either direct visual inspection or by a neural-network–inspired algorithm called M2-net."},{"metadata":{},"cell_type":"markdown","source":"## [Baidu Algorithms](https://syncedreview.com/2020/05/12/new-baidu-algorithms-boost-mrna-vaccine-development-for-sars-cov-2/)"},{"metadata":{},"cell_type":"markdown","source":"“The LinearDesign algorithm, developed by Baidu Research in collaboration with Oregon State University and University of Rochester, can theoretically design the mRNA sequence with the most stable secondary structure, helping many mRNA vaccine companies to optimize their vaccine sequence designs,” says Liang Huang, Distinguished Scientist at Baidu USA."},{"metadata":{},"cell_type":"markdown","source":"It’s know that mRNA vaccines may fail due to degradation during storage and transportation and resultant low protein levels. "},{"metadata":{},"cell_type":"markdown","source":"In a recent paper published on Proceedings of the National Academy of Sciences in the US, Moderna’s research team demonstrated that secondary structures and codon optimality can increase mRNA stability and protein expression. The problem can therefore be formulated to finding mRNA sequences that are good in both secondary structure stability and codon optimality among the exponentially many synonymous sequences that encode the same protein. This is undoubtedly difficult."},{"metadata":{},"cell_type":"markdown","source":"To find an optimal mRNA sequence, scientists have traditionally made random changes to a sequence and then seen if they were beneficial. The scientific community is now seeking different approaches to solve the problem. For example, Eterna, a browser-based gaming platform developed by Stanford University, is assembling online gamers to develop a safe mRNA vaccine by solving puzzles. Eterna has been using Baidu’s LinearFold algorithm to accelerate secondary structure analysis."},{"metadata":{},"cell_type":"markdown","source":"## [Ref cited in Prev link](https://www.pnas.org/content/116/48/24075.short?rss=1)"},{"metadata":{},"cell_type":"markdown","source":"Despite widespread recognition that RNA is inherently structured, the interplay between local and global mRNA secondary structure (particularly in the coding region) and overall protein expression has not been thoroughly explored. Our work uses 2 approaches to disentangle the regulatory roles of mRNA primary sequence and secondary structure: global substitution with modified nucleotides and computational sequence design. By fitting detailed kinetic expression data to mathematical models, we show that secondary structure can increase mRNA half-life independent of codon usage. These findings have significant implications for both translational regulation of endogenous mRNAs and the emerging field of mRNA therapeutics."},{"metadata":{},"cell_type":"markdown","source":"## [Dilated Convolutions](https://ankitg.me/blog/2017/10/03/dilated-convolutions.html)"},{"metadata":{},"cell_type":"markdown","source":"This “linear” strand is actually tightly compacted in 3-dimensional space, meaning that DNA locations that are far apart along the 1-dimensional sequence may be adjacent and able to interact. The nature of this compacting is thought to be regulated by a variety of factors, including chemical modifications to bound proteins called histones and the binding of external proteins called transcription factors."},{"metadata":{},"cell_type":"markdown","source":"Thus, if we are trying to predict the locations of regulatory markers directly from a DNA sequence, it follows that capturing information across a very wide strand of DNA should improve accuracy. In machine learning, we formally refer to this property as receptive field size. The receptive field of an output neuron in a neural network is the set of inputs that affects it. Thus, we wish to build a model with large receptive fields. In order to naively scale existing approaches for regulatory marker prediction to larger receptive fields, the models would require a linear increase in the number of parameters, which makes them prone to overfitting and reduces training speed. Moreover, usually effective recurrent neural network models such as LSTM-based networks are difficult to train due to the long sequence lengths."},{"metadata":{},"cell_type":"markdown","source":"To address this challenge, we need a model in which the number of parameters grows sublinearly with respect to the input size, and in which we can parallelize our predictions across the input sequence (as it may be very long). The first constraint makes standard convolutional neural networks impractical, as their parameters grow linearly with respect to the receptive field of a given output. The latter constraint makes standard LSTM-based networks impractical too."},{"metadata":{},"cell_type":"markdown","source":"Instead, we use a dilated convolutional neural network. The receptive field size of a dilated CNN is exponential in the number of layers. This means that we can scale our models to several orders of magnitude more inputs than existing approaches, while only requiring a small increase in the number of parameters. Moreover, since these are adaptations of convolutions, we can implement these in a highly-parallelizable manner on GPU hardware in standard neural network packages. "},{"metadata":{},"cell_type":"markdown","source":"## [Unrolled Algorithms](https://openreview.net/forum?id=S1eALyrYDH)"},{"metadata":{},"cell_type":"markdown","source":"We propose an end-to-end deep learning model, called E2Efold, for RNA secondary structure prediction which can effectively take into account the inherent constraints in the problem. The key idea of E2Efold is to directly predict the RNA base-pairing matrix, and use an unrolled algorithm for constrained programming as the template for deep architectures to enforce constraints."},{"metadata":{},"cell_type":"markdown","source":"## [Folding](https://grantome.com/grant/NIH/R01-GM058545-15)"},{"metadata":{},"cell_type":"markdown","source":"The folding of RNA molecules into specific tertiary structures is important for many cellular processes relating to gene expression"},{"metadata":{},"cell_type":"markdown","source":"## [Dynamics and Stability of Individual Base Pairs](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2758270/)"},{"metadata":{},"cell_type":"markdown","source":"within the same base sequence context, dA-rU base pairs are less stable than dT-rA base pairs. These differences in stability are largest when two dA-rU base pairs are located next to each other in the structure"},{"metadata":{},"cell_type":"markdown","source":"In these previous studies, the largest differences in stability were observed for RNA-DNA hybrids that contained all purine (R) and all pyrimidine (Y) bases in each strand. Specifically, the hybrids rR-dY (i.e., purines in the RNA strand and pyrimidines in the DNA strand) are more stable than the homologous dR-rY hybrids"},{"metadata":{},"cell_type":"markdown","source":"These findings offer important insights into the structural origin of the differences in stability between dA-rU and rA-dT base pairs observed in the present work. They suggest that rA-dT base pairs are stabilized relative to dA-rU base pairs due to the presence of the C5 methyl group in thymine and to the location of the 2’-OH group in the purine."},{"metadata":{},"cell_type":"markdown","source":"## [Stem-loops](https://www.sciencedirect.com/science/article/abs/pii/S0022519398906749)"},{"metadata":{},"cell_type":"markdown","source":"These are the quantities of the four bases themselves, and six derived parameters: ATmin, CGmin, Chargaff differences for the W and S bases, and two base products. The quantities of the least represented base of a Watson–Crick base pair (ATmin, CGmin) might provide an index of the total stem potential of a window. The degrees to which one base of a Watson–Crick pair exceeds the other (the Chargaff differences for the W bases and for the S bases) might provide an index of the total loop potential of a window. Base products (A×T, C×G) might provide an index both of stem and of loop potentials."},{"metadata":{},"cell_type":"markdown","source":"## [DNA stacking](https://academic.oup.com/nar/article/34/2/564/2401647)"},{"metadata":{},"cell_type":"markdown","source":"Two factors are mainly responsible for the stability of the DNA double helix: base pairing between complementary strands and stacking between adjacent bases. "},{"metadata":{},"cell_type":"markdown","source":"From stacking parameters of individual contacts, we calculate base-stacking contribution to the stability of A•T- and G•C-containing DNA polymers. We find that temperature and salt dependences of the stacking term fully determine the temperature and the salt dependence of DNA stability parameters. For all temperatures and salt concentrations employed in present study, base-stacking is the main stabilizing factor in the DNA double helix. A•T pairing is always destabilizing and G•C pairing contributes almost no stabilization. "},{"metadata":{},"cell_type":"markdown","source":"## [Predictions](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005827)"},{"metadata":{},"cell_type":"markdown","source":"Predictions of RNA tertiary structure tend to be least accurate in loop regions, where non-canonical pairs are important for determining the details of structure."},{"metadata":{},"cell_type":"markdown","source":"In general, automated structure prediction methods can identify the global fold of an RNA, but modeling of loop regions is frequently inaccurate. These poorly-predicted regions are often highly ordered, containing specific arrangements of non-canonical base pairs "},{"metadata":{},"cell_type":"markdown","source":"The canonical RNA secondary structure (the set of A-U, G-C, and G-U wobble pairs in an RNA structure) can be predicted using a free energy change model developed from optical melting experiments on short RNA sequences combined with a dynamic programming algorithm that efficiently searches the space of possible structures to identify the best structure according to the model. This method implicitly considers the free energy change of some non-canonical pairs, such as mismatched nucleotides at the end of helices, but it does not explicitly make predictions of non-canonical pairs."},{"metadata":{},"cell_type":"markdown","source":"## [BPP](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-018-2018-4)"},{"metadata":{},"cell_type":"markdown","source":" the base-pairing probabilities (BPPs),, which are often greater than eighty percent, are convenient to observe the local stability of the secondary structures. Furthermore, the estimators based on maximum expected accuracy, such as the MEA (Maximum Expected Accuracy) estimator and the γ-centroid estimator, can be calculated by posterior decoding of BPPs without using further information of the RNA sequence"},{"metadata":{},"cell_type":"markdown","source":"## [Main Source](https://www.frontiersin.org/articles/10.3389/fgene.2019.00467/full)"},{"metadata":{},"cell_type":"markdown","source":"We built the matrix Wi×i for each RNA, where each row of the matrix represents possible pairings of bases at that position, as follows:\n\n1. According to the number of hydrogen bonds between the paired bases, the pairing weight between A and U is set to 2, and the pairing weight between G and C is set to 3. Since the U-G pair is a wobble base pair, the pairing weight between U and G is set to x (0 < x < 2), which leads to:...\n\n2. For any two positions, such as i, j on the RNA sequence, this article must not only consider the pairing of these two bases, but also whether these two positions can form paired bases on the stem. Therefore, we had to take into account the pairing of the bases on the left (right) side of i with the right (left) side of j.\n\n3. For one stem, the paired bases in the middle of the stem are relatively stable, and the paired bases on both sides are relatively unstable. Therefore, the calculation offers the possibility of pairing the two positions, i and j, on an RNA sequence. This article refers to the idea of the local weighted linear regression, adding a Gaussian function as a weight. The closer i and j were, the higher the weight of the paired bases and the greater their effect."},{"metadata":{},"cell_type":"markdown","source":"## [ML](https://www.biorxiv.org/content/10.1101/303172v1.full.pdf)"},{"metadata":{},"cell_type":"markdown","source":"In  this  paper,  we  propose  a  novel  algorithm  to  directly  infer  base-pairing  probabilities  with  neural  networks  instead  of  the  McCaskillalgorithm   or   the   inside-outside   algorithm,    which   depend   on   thearchitecture of RNA secondary structures"},{"metadata":{},"cell_type":"markdown","source":"## EOF"}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}