{"cells":[{"metadata":{"_uuid":"8f2839f25d086af736a60e9eeb907d3b93b6e0e5","_cell_guid":"b1076dfc-b9ad-4769-8c92-a6c4dae69d19","trusted":true},"cell_type":"markdown","source":"<img src='https://img.timesnownews.com/story/1536770343-miedicine-2.jpg?d=600x450'>\n<p>\n<h1><center>Mechanisms of Action(MoA) Prediction: EDA!üíä</center><h1>\n    \n    \n    \n# 1. <a id='Introduction'>Introduction üÉè </a>\n    \n### 1.1 What is Mechanisms of Action(MoA)?\n* In the past, scientists derived drugs from natural products or were inspired by traditional remedies. Very common drugs, such as paracetamol, known in the US as acetaminophen, were put into clinical use decades before the biological mechanisms driving their pharmacological activities were understood. Today, with the advent of more powerful technologies, drug discovery has changed from the serendipitous approaches of the past to a more targeted model based on an understanding of the underlying biological mechanism of a disease. In this new framework, scientists seek to identify a protein target associated with a disease and develop a molecule that can modulate that protein target. As a shorthand to describe the biological activity of a given molecule, scientists assign a label referred to as mechanism-of-action or MoA for short.\n\n###  1.2 What is Mechanisms of Action(MoA) Prediction Competition?\n* The aim of this challenge is to ‚Äúclassify drugs based on their biological activity‚Äù. Pharmaceutical drug discovery aims to identify certain proteins that are associated with a specific disease, and then to develop molecules that can target those proteins. The MoA of a molecule encodes its biological activity. Our dataset describes the response of 100 different types of human cells to various drugs. Those response patterns will be used to classify the MoA reponse.\n(This is a multi-label classification problem.)  \n\n### 1.3 Metric: log loss\n* Please see the [evaluation metric](http://https://www.kaggle.com/c/lish-moa/overview/evaluation).","execution_count":null},{"metadata":{"_uuid":"d629ff2d2480ee46fbb7e2d37f6b5fab8052498a","_cell_guid":"79c7e3d0-c299-4dcb-8224-4455121ee9b0","trusted":true},"cell_type":"code","source":"import os\n\nimport numpy as np\nimport pandas as pd\nimport matplotlib.pyplot as plt\nimport seaborn as sns\n\n#color\nfrom colorama import Fore, Back, Style","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# 2. <a id='2'>Reading the data üìö</a>","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"base_path = '../input/lish-moa/'","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"* train_features.csv - Features for the training set. Features g- signify gene expression data, and c- signify cell viability data. cp_type indicates samples treated with a compound (cp_vehicle) or with a control perturbation (ctrl_vehicle); control perturbations have no MoAs; cp_time and cp_dose indicate treatment duration (24, 48, 72 hours) and dose (high or low).\n* train_targets_scored.csv - The binary MoA targets that are scored.\n* train_targets_nonscored.csv - Additional (optional) binary MoA responses for the training data. These are not predicted nor scored.\n* test_features.csv - Features for the test data. You must predict the probability of each scored MoA for each row in the test data.\n* sample_submission.csv - A submission file in the correct format.","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"test_features_df = pd.read_csv(base_path + 'test_features.csv')\ntrain_features_df = pd.read_csv(base_path + 'train_features.csv')\ntrain_targets_scored_df = pd.read_csv(base_path + 'train_targets_scored.csv')\ntrain_targets_nonscored_df = pd.read_csv(base_path + 'train_targets_nonscored.csv')\nsample_submission_df = pd.read_csv(base_path + 'sample_submission.csv')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"print(Fore.YELLOW + 'Sample submission shape: ',Style.RESET_ALL,sample_submission_df.shape)\nsample_submission_df.head(5)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# 3. <a id='3'>Basic Data Exploration üèïÔ∏è</a>","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"# Null values and Data types\nprint(Fore.YELLOW + 'test_features_df !!',Style.RESET_ALL)\nprint(test_features_df.info())\nprint('-------------')\nprint(Fore.BLUE + 'train_features_df !!',Style.RESET_ALL)\nprint(train_features_df.info())\nprint(Fore.YELLOW + 'train_targets_scored_df !!',Style.RESET_ALL)\nprint(train_targets_scored_df.info())\nprint('-------------')\nprint(Fore.BLUE + 'train_targets_nonscored_df !!',Style.RESET_ALL)\nprint(train_targets_nonscored_df.info())\nprint(Fore.YELLOW + 'sample_submission_df !!',Style.RESET_ALL)\nprint(test_features_df.info())","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"float64(`872`) : 772 columns with gene expression data + 100 columns with cell viability data","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"## Correlation\n* https://www.kaggle.com/blessondensil294/beginners-eda-viz-moa-drug-prediction","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"corrmat = train_features_df.corr()\nf, ax = plt.subplots(figsize=(14,14))\nsns.heatmap(corrmat, square=True, vmax=.8)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"corrmat = train_targets_scored_df.corr()\nf, ax = plt.subplots(figsize=(14,14))\nsns.heatmap(corrmat, square=True, vmax=.8)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"corrmat = train_targets_nonscored_df.corr()\nf, ax = plt.subplots(figsize=(14,14))\nsns.heatmap(corrmat, square=True, vmax=.8)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## Automated ML tools for EDA\n\nFor basic EDA, I'll use some automated ML tools.","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"# 4. <a id='4'>Data Analysis Baseline Library (dabl) üå±</a> ","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"For basic EDA, I'll use some automated ML tools.","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"This project tries to help make supervised machine learning more accessible for beginners, and reduce boiler plate for common tasks","execution_count":null},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"!pip install dabl","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import dabl","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"Now let‚Äôs ask dabl what it thinks by cleaning up the data.\n\ndabl tries to detect the types of your data and apply appropriate conversions. It also tries to detect potential data quality issues. The field of data cleaning is impossibly broad, and dabl‚Äôs approaches are by no means sophisticated. The goal of dabl is to get the data ‚Äúclean enough‚Äù to create useful visualizations and models, and to allow users to perform custom cleaning operations themselves. In particular if the detection of semantic types (continuous, categorical, ordinal, text, etc) fails, the user can provide type_hints:","execution_count":null},{"metadata":{"trusted":true,"_kg_hide-output":true},"cell_type":"code","source":"test_features_df_clean = dabl.clean(test_features_df, verbose=0)\ntrain_features_df_clean = dabl.clean(train_features_df, verbose=0)\ntrain_targets_scored_df_clean = dabl.clean(train_targets_scored_df, verbose=0)\ntrain_targets_nonscored_df_clean = dabl.clean(train_targets_nonscored_df, verbose=0)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# EDA for test_features_df","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"types = dabl.detect_types(test_features_df_clean)\nprint(types) ","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### cp_type","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"`cp_type` indicates samples treated with a compound (cp_vehicle) or with a control perturbation (ctrl_vehicle).","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(test_features_df_clean, 'cp_type')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### cp_time","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"`cp_time` : treatment duration (24, 48, 72 hours).","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(test_features_df_clean, 'cp_time')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### cp_dose","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"`cp_dose` : dose (high or low)","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(test_features_df_clean, 'cp_dose')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### g-0 ~ g-771","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"Features `g-` signify gene expression data. Gene expression is the process by which the information encoded in a gene is used to direct the assembly of a protein molecule. The cell reads the sequence of the gene in groups of three bases.\n\nIn test_feature_df, we can see `g-0` ~ `g-771` and I'll show `g-0, g-100, g-350, g-500, g-771` since there are too many columns for `g-`.","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(test_features_df_clean, 'g-0')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(test_features_df_clean, 'g-100')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(test_features_df_clean, 'g-350')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(test_features_df_clean, 'g-500')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(test_features_df_clean, 'g-771')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### C-0 ~ C-99","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"`c-` signify `cell viability` data. Cell viability is a measure of the proportion of live, healthy cells within a population. Cell viability assays are used to determine the overall health of cells, optimize culture or experimental conditions, and to measure cell survival following treatment with compounds, such as during a drug screen.\n\nIn test_feature_df, we can see `c-0` ~ `c-99` and I'll show `c-0, c-50, c-99` since there are too many columns for `c-`.","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(test_features_df_clean, 'c-0')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(test_features_df_clean, 'c-50')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(test_features_df_clean, 'c-99')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# EDA for train_features_df","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"types = dabl.detect_types(train_features_df_clean)\nprint(types) ","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### cp_type","execution_count":null},{"metadata":{"trusted":true,"_kg_hide-output":true},"cell_type":"code","source":"dabl.plot(train_features_df_clean, 'cp_type')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### cp_time","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(train_features_df_clean, 'cp_time')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### cp_dose","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(train_features_df_clean, 'cp_dose')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"### g-0 ~ g-771\nIn train_features_df, we can see `g-0` ~ `g-771` and I'll show `g-0, g-100, g-350, g-500, g-771` since there are toomany columns for `g-`.","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(train_features_df_clean, 'g-0')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(train_features_df_clean, 'g-100')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(train_features_df_clean, 'g-350')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(train_features_df_clean, 'g-500')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(train_features_df_clean, 'g-771')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"In train_features_df, we can see `c-0` ~ `c-99` and I'll show `c-0, c-50, c-99` since there are too many columns for `c-`.","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(train_features_df_clean, 'c-0')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(train_features_df_clean, 'c-50')","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"dabl.plot(train_features_df_clean, 'c-99')","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# EDA for train_targets_scored_df","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"types = dabl.detect_types(train_targets_scored_df_clean)\nprint(types) ","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"train_targets_scored_df_clean","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"train_targets_scored_df","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"All is `0`.","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"# EDA for train_targets_nonscored_df","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"train_targets_nonscored_df_clean","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"train_targets_nonscored_df","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"All is `0`.","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"# 5. <a id='5'>Datasist ‚ú®</a> ","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"![](https://warehouse-camo.ingress.cmh1.psfhosted.org/6572d848c045b008268a4d6ca2617526a102d9b0/68747470733a2f2f726973656e772e6769746875622e696f2f64617461736973742f64617461736973742e706e67)\n**datasist** is a python package providing fast, quick, and an abstracted interface to popular and frequently used functions or techniques relating to data analysis, visualization, data exploration, feature engineering, Computer, NLP, Deep Learning, modeling, model deployment etc.","execution_count":null},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"!pip install datasist","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import datasist as ds  #import datasist library","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**check_train_test_set**: Checks the distribution of train and test for uniqueness in order to determine the best feature engineering strategy.","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"ds.structdata.check_train_test_set(train_features_df, test_features_df, index=None, col=None)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"ds.structdata.describe(test_features_df)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"ds.structdata.describe(train_features_df)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**detect_outliers**: Detect Rows with outliers.","execution_count":null},{"metadata":{"trusted":true,"_kg_hide-output":true},"cell_type":"code","source":"numerical_feats = ds.structdata.get_num_feats(test_features_df)\nds.structdata.detect_outliers(test_features_df,80,numerical_feats)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true,"_kg_hide-output":true},"cell_type":"code","source":"numerical_feats = ds.structdata.get_num_feats(train_features_df)\nds.structdata.detect_outliers(train_features_df,80,numerical_feats)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**display_missing**: Display missing values as a pandas dataframe.","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"ds.structdata.display_missing(test_features_df)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"ds.structdata.display_missing(train_features_df)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**get_cat_feats** : Returns the categorical features in a data set","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"cat_feats = ds.structdata.get_cat_feats(test_features_df)\ncat_feats","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"cat_feats = ds.structdata.get_cat_feats(train_features_df)\ncat_feats","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**get_num_feats** : Returns the numerical features in a data set","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"num_feats = ds.structdata.get_num_feats(test_features_df)\nprint(len(num_feats))","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"num_feats = ds.structdata.get_num_feats(train_features_df)\nprint(len(num_feats))","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"get_unique_counts = ds.structdata.get_unique_counts(test_features_df)\nget_unique_counts","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"get_unique_counts = ds.structdata.get_unique_counts(train_features_df)\nget_unique_counts","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"# 6. <a id='6'>AutoViz üõ∂ </a> ","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"![](https://github.com/AutoViML/AutoViz/raw/master/logo.png)\nAutomatically Visualize any dataset, any size with a single line of code.\n\nAutoViz performs automatic visualization of any dataset with one line. Give any input file (CSV, txt or json) and AutoViz will visualize it.","execution_count":null},{"metadata":{"trusted":true,"_kg_hide-output":true},"cell_type":"code","source":"!pip install autoviz","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import pandas as pd\nfrom autoviz.AutoViz_Class import AutoViz_Class\nAV = AutoViz_Class()","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## test_features","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"test_features_df","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"sep = ','\ntarget = 'cp_type'\ndft = AV.AutoViz(filename=\"\", sep=sep, depVar=target, dfte=test_features_df, header=0, verbose=1,\n                            lowess=False,chart_format='svg',max_rows_analyzed=5000,max_cols_analyzed=50)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"sep = ','\ntarget = 'cp_type'\ndft = AV.AutoViz(filename=\"\", sep=sep, depVar=target, dfte=train_features_df, header=0, verbose=1,\n                            lowess=False, chart_format='svg',max_rows_analyzed=4000,max_cols_analyzed=50)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"I can not get Plot for data.csv, but get some information.\n* Number of variables removed due to `high correlation` = `235` in test_features_df\n* Number of variables removed due to `high correlation` = `229`  in train_features_df","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"# 7. <a id='7'>missingno üõ∂</a> ","execution_count":null},{"metadata":{},"cell_type":"markdown","source":"![](https://storage.googleapis.com/coderzcolumn/static/tutorials/data_science/article_image/missingno%20-%20Visualize%20Missing%20Data%20in%20Python.jpg)\nMessy datasets? Missing values? missingno provides a small toolset of flexible and easy-to-use missing data visualizations and utilities that allows you to get a quick visual summary of the completeness (or lack thereof) of your dataset. Just pip install missingno to get started.\n\nIn the case of a real-world dataset, it is very common that some values in the dataset are missing. We represent these missing values as NaN (Not a Number) values. But to build a good machine learning model our dataset should be complete. That‚Äôs why we use some imputation techniques to replace the NaN values with some probable values. But before doing that we need to have a good understanding of how the NaN values are distributed in our dataset.\n\nMissingno library offers a very nice way to visualize the distribution of NaN values. Missingno is a Python library and compatible with Pandas.","execution_count":null},{"metadata":{"_kg_hide-output":true,"trusted":true},"cell_type":"code","source":"!pip install missingno","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"import missingno as msno\n\nimport matplotlib.pyplot as plt\n%matplotlib inline","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**Matrix:**\n\nVisualising missing values for a sample. Using this matrix you can very quickly find the pattern of missingness in the dataset.","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"msno.matrix(test_features_df)","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"msno.matrix(train_features_df)","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"**Bar Chart :**\n\nThis bar chart gives you an idea about how many missing values are there in each column.","execution_count":null},{"metadata":{"trusted":true},"cell_type":"code","source":"msno.bar(test_features_df.sample(10))","execution_count":null,"outputs":[]},{"metadata":{"trusted":true},"cell_type":"code","source":"msno.bar(train_features_df.sample(10))","execution_count":null,"outputs":[]},{"metadata":{},"cell_type":"markdown","source":"## If this kernel is useful, <font color='orange'>please upvote</font>!\n- See you next time and I will update it soon!","execution_count":null}],"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat":4,"nbformat_minor":4}